---
title: "encoded_access_test"
output: html_document
---


```{r}
#minor change
#library(Seurat)
library(jsonlite)
library(httr)
library(stringr)
library(data.table)
library(doParallel)
registerDoParallel(detectCores()-1)
```



```{r, create directory structure for project}
dirs <- list(data_dir = "./data",
             results_dir = "./results",
             figures_dir_parent = "./results/figures",
             figures_dir = "./results/figures/raw",
             tables_dir = "./results/tables")
dirs$data_dir <- "/u/scratch/l/leroybon/chrom_mod_proj_data"
for(d in dirs){
  if(!dir.exists(d)){dir.create(d)}
}
date <- format(Sys.time(),"%Y%m%d")
```


```{r}
dataset_list <- list()
```

```{r, read early development}
## zygote, 2 cell, 4 cell, 8 cell, 16 cell, blasts cell

dirs$GSE45719 <- paste0(dirs$data_dir,"/GSE45719")
if(!(dir.exists(dirs$GSE45719))){
  dir.create(dirs$GSE45719)
  system(paste0("curl ",
                "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE45nnn/",
                "GSE45719/suppl/GSE45719_RAW.tar",
                paste0(" > ",dirs$GSE45719,"/GSE45719_RAW.tar" )))
  system(paste0("tar -xf ",dirs$GSE45719,"/GSE45719_RAW.tar -C ", dirs$GSE45719))
  
  files <- list.files(dirs$GSE45719)[grep(".txt.gz", list.files(dirs$GSE45719))]
  for(f in files){
   system(paste0("gunzip ", dirs$GSE45719,"/",f )) 
  }
  temp_list <- list()
  files <- list.files(dirs$GSE45719, full.names = T)[grep("expression.txt", list.files(dirs$GSE45719))]
  temp_df <- str_split_fixed(str_split_fixed(files,"GSE45719/|_expression",3)[,2], "_|-", 5)[,1:4]
  colnames(temp_df) <- c("GSM_ID","stage","rep","cell_id")
  files_df <- data.frame(file=files, temp_df)
  files_df <- files_df[!grepl("BXC|fibroblast|pooled|split|smartseq",files_df$file),]
  for(i in 1:nrow(files_df)){
   temp_df <- fread(files_df$file[i])
   temp_df <- temp_df[,c("#Gene_symbol","reads")]; colnames(temp_df)[1] <- "gene_symbol"
   temp_df <- aggregate(reads ~ gene_symbol, temp_df,sum)
   temp_list[[paste(files_df[i,3:ncol(files_df)], collapse="_")]] <- temp_df
  }
  temp_df <- as.data.frame(matrix(nrow=length(temp_list[[1]]$gene_symbol), ncol=length(temp_list)))
  rownames(temp_df) <- temp_list[[1]]$gene_symbol; colnames(temp_df) <- names(temp_list)
  for(i in 1:length(temp_list)){
    if(any(temp_list[[i]]$gene_symbol != rownames(temp_df))){stop("gene names do not match")}
    temp_df[,i] <- temp_list[[i]]$reads
  }
}
dataset_list$GSE45719 <- list(mat=temp_df,meta=NA)
```

```{r, read early gastrution paper ref data}
## e3.5, e4.5, e5.5, e6.5
target <- "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE100nnn/GSE100597/suppl/GSE100597%5Fcount%5Ftable%5FQC%5Ffiltered%2Etxt%2Egz"
temp <- tempfile()
download.file(target, temp)
temp_data <- read.table(temp, skip = 0, header = TRUE, sep = "\t")
rownames(temp_data) <- temp_data$X
temp_data$X <- NULL
temp_data[,1:ncol(temp_data)] <- apply(temp_data,2,as.numeric)

dataset_list[["GSE100597"]] <- list(mat=temp_data, meta=NA)
```

```{r, read early organogenesis paper ref data}
## Note this file is fairly large ~10 Gb
## e6.5, 7.25, e7.75, 8.5
library(data.table)
library(Seurat)

dirs$e_mtab_6967_atlas <- paste0(dirs$data_dir, "/e_mtab_6967_atlas")
if(!dir.exists(dirs$e_mtab_6967_atlas)){
  dir.create(dirs$e_mtab_6967_atlas)
  system(paste0("curl ",
                "https://www.ebi.ac.uk/arrayexpress/files/",
                "E-MTAB-6967/E-MTAB-6967.processed.1.zip ",
                "> ",dirs$e_mtab_6967_atlas ,"/atlas_data.zip"))
  system(paste0("unzip ",dirs$e_mtab_6967_atlas,"/atlas_data.zip -d ",dirs$e_mtab_6967_atlas))
  system(paste0("rm ",dirs$e_mtab_6967_atlas,"/atlas_data.zip"))
  system(paste0("tar -zxvf ",dirs$e_mtab_6967_atlas,"/atlas_data.tar.gz -C ",dirs$e_mtab_6967_atlas))
  system(paste0("mv ",dirs$e_mtab_6967_atlas,"/atlas",dirs$e_mtab_6967_atlas,"/e_mtab_6967_atlas"))
}

temp_data <- ReadMtx( paste0(dirs$e_mtab_6967_atlas,"/atlas/raw_counts.mtx"),
                     cells = paste0(dirs$e_mtab_6967_atlas,"/atlas/meta.csv"),
                     features = paste0(dirs$e_mtab_6967_atlas,"/atlas/genes.tsv"),
                     skip.cell = 1,
                     cell.sep = ",")

cell_meta <- fread(paste0(dirs$e_mtab_6967_atlas,"/atlas/meta.csv"))

which <- which(!cell_meta$doublet)
temp_data <- temp_data[,which]
cell_meta <- cell_meta[which,]

entities <- apply(cell_meta[,c("stage","cluster","cluster.sub","celltype","haem_subclust","endo_gutCluster")],1,paste,collapse=",")
tempdf <- unique(cell_meta[,c("stage","cluster","cluster.sub","celltype","haem_subclust","endo_gutCluster")])
entity_types <- unique(entities)
cell_entity_map <- list()
for(i in 1:length(entity_types)){
  which <- which(entities == entity_types[i])
  cell_entity_map[[entity_types[i]]] <- cell_meta$cell[which]
}

df <- data.frame(matrix(nrow = nrow(temp_data), ncol=length(entity_types)))
rownames(df) <- rownames(temp_data)
colnames(df) <- names(cell_entity_map)

for(i in 1:ncol(df)){
  if(i %% 50 ==0){print(i)}
  df[,i] <- rowSums(temp_data[,cell_entity_map[[i]],drop=F])
}

for(i in 1:ncol(df)){
  tempdf$num_cells[i] <- length(cell_entity_map[[i]])
}

dataset_list$E_MTAB_6967 <- list(mat=df,meta=tempdf) 
rm(temp_data)
```



```{r, read mouse organogenesis paper}
# E9.5 to E13.5

dirs$GSE119945 <- paste0(dirs$data_dir,"/GSE119945/")
if(!(dir.exists(dirs$GSE119945))){
  dir.create(dirs$GSE119945)
  system(paste0("curl ",
                "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE119nnn/",
                "GSE119945/suppl/GSE119945%5Fgene%5Fcount%2Etxt%2Egz",
                paste0("> ",dirs$GSE119945,"gene_count.txt.gz" )))
  system(paste0("gunzip -c ",dirs$GSE119945,"gene_count.txt.gz > ",dirs$GSE119945,"gene_count.txt"))
  system(paste0("curl ",
                "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE119nnn/",
                "GSE119945/suppl/GSE119945%5Fcell%5Fannotate%2Ecsv%2Egz",
                paste0("> ",dirs$GSE119945,"cell_annotate.csv.gz" )))
  system(paste0("curl ",
                "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE119nnn/",
                "GSE119945/suppl/GSE119945%5Fgene%5Fannotate%2Ecsv%2Egz",
                paste0("> ",dirs$GSE119945,"gene_annotate.csv.gz" )))
}

temp_cells <- fread(paste0(dirs$GSE119945,"cell_annotate.csv.gz" ))
temp_cells <- as.data.frame(temp_cells)
temp_cells$Sub_Cluster[grep("unknown", temp_cells$Sub_Cluster)] <- "0"
temp_features <- fread(paste0(dirs$GSE119945,"gene_annotate.csv.gz" ))
temp_features <- as.data.frame(temp_features)
temp_cells$cluster <- apply(temp_cells[,c("Main_Cluster","Sub_Cluster")],
                            1,paste,collapse="-")
temp_cluster_exp_mat <- matrix(0,nrow = nrow(temp_features),
                          ncol = length(unique(temp_cells$cluster)) )
rownames(temp_cluster_exp_mat) <- temp_features$gene_id
colnames(temp_cluster_exp_mat) <- unique(temp_cells$cluster)
## num lines gene_count.txt.gz is at max cells x features + 2
temp_nrow <- 500e6 
temp_num <- ceiling(file.size(paste0(dirs$GSE119945,"gene_count.txt")) / temp_nrow)

dirs$temp_split <- paste0(dirs$GSE119945,"temp_split")
if(!dir.exists(dirs$temp_split)){
  dir.create(dirs$temp_split)
  system(paste0("split ",dirs$GSE119945,"gene_count.txt",
                " -n ", "l/",temp_num, " ",
                paste0(dirs$temp_split,"/chunk_")
                )
  )
}


file_list <- list.files(dirs$temp_split, full.names = T) 
file_list <- file_list[grep("chunk_", file_list)]

dataset_list$GSE119945$mat <- foreach( f=file_list, .combine = "+") %dopar% {
  if(grepl("chunk_aa",f)){
     # temp_header <- system(paste0("head -n 2 ", f), intern = T)
     # fileConn <- file(paste0(dirs$GSE119945,"temp_split/","header.txt"))
     # writeLines(temp_header, fileConn)
     # close(fileConn)
     mat_part <- fread(f, skip = 2, header = F,
                    col.names = c("gene","cell","value") )
   } else {
     mat_part <- fread(f, skip = 0, header = F,
                    col.names = c("gene","cell","value") )
   }
  mat_part <- as.data.frame(mat_part)
  mat_part$cluster <- temp_cells$cluster[match(mat_part$cell, 1:nrow(temp_cells))]
  ## next need to aggregate counts for each gene for each cluster ....
  
  
  # temp_mp<- aggregate(value ~ gene + cluster, data = mat_part, FUN=sum)
  
  temp_clusters <- unique(mat_part$cluster)
  for(i in 1:length(temp_clusters)){
    mat_part_sub <- mat_part[which(mat_part$cluster==temp_clusters[i]),]
    temp_mp <- aggregate(value ~ gene + cluster, data = mat_part_sub, FUN=sum)
    temp_cluster_exp_mat[temp_mp$gene, unique(temp_mp$cluster)] <-
      temp_cluster_exp_mat[temp_mp$gene, unique(temp_mp$cluster)] + temp_mp$value
  }
  temp_cluster_exp_mat
}

dataset_list$GSE119945$meta <- temp_cells
rm(temp_cells)
```


```{r, read encode mouse ref RNAseq}

## e10.5, e11.5, e12.5, e13.5, e14.5, e15.5, e16.5, neonat, 8wk adult, 24wk adult
## get ENCODE search results : mouse epigenome reference RNAseq
search_query <- "https://www.encodeproject.org/search/?type=Experiment&related_series.@type=ReferenceEpigenome&status=released&replicates.library.biosample.donor.organism.scientific_name=Mus+musculus&assay_title=polyA+plus+RNA-seq&frame=object&format=json&limit=all"
query_result <- GET(search_query)
query_result <- content(query_result , "text", encoding = "UTF-8")
query_result <- fromJSON(query_result, flatten = T)
query_table <- query_result$`@graph`[!unlist(lapply(query_result$`@graph`, is.list))]

##
data_list <- list()
for(i in 1:nrow(query_table)){
  if(i %% 5 == 0){print(i)}
  target <- paste0("https://www.encodeproject.org/",
                          query_table$accession[i],
                          "/?frame=embedded&format=json")
  target <- GET(target)
  target_result <- content(target, "text", encoding = "UTF-8")
  target_result <- fromJSON(target_result, flatten = TRUE)
  temp_gene_anno <-  unique(target_result$files$genome_annotation)
  temp_gene_anno <- temp_gene_anno[grep("ENSEMBL", temp_gene_anno,invert = T)]
  which <- which(target_result$files$file_format=="tsv" &
                 target_result$files$output_type=="gene quantifications" &
                 grepl(max(as.numeric(str_remove_all(temp_gene_anno,"[a-zA-Z]")),na.rm = T),
                       target_result$files$genome_annotation) &
                 target_result$files$status == "released")
  which_cols <- c("accession", "href","biosample_ontology",
                "biological_replicates", "technical_replicates",
                "biological_replicates_formatted", "donors",
                "dataset")
  gene_quant_files <- target_result$files[which,which_cols]
  gene_quant_files[,which(unlist(lapply(gene_quant_files, is.list)))] <- 
    sapply(gene_quant_files[,which(unlist(lapply(gene_quant_files, is.list)))], paste )
  target_metadata <- cbind(query_table[i,], gene_quant_files)
  
  for(j in 1:nrow(target_metadata)){
    temp <- tempfile()
    download.file(paste0("https://www.encodeproject.org/",target_metadata$href[j]), temp)
    temp_data <- read.table(temp, skip = 0, header = TRUE, sep = "\t")
    data_list[[length(data_list)+1]] <- list(TPM = data.frame(TPM=temp_data$TPM),
                                             metadata = cbind(target_metadata[j,], read_count=sum(temp_data$posterior_mean_count) ) )
    rownames(data_list[[length(data_list)]]$TPM) <- temp_data$gene_id
  }
}

temp_data_list_qc <- data.frame(tpm_nrow=numeric(length=length(data_list)),
                                tpm_ncol=numeric(length=length(data_list)),
                                metadata_nrow=numeric(length=length(data_list)))
for(i in 1:length(data_list)){
  temp_data_list_qc$tpm_nrow[i] <- nrow(data_list[[i]]$TPM)
  temp_data_list_qc$tpm_ncol[i] <- ncol(data_list[[i]]$TPM)
  temp_data_list_qc$metadata_nrow[i] <- nrow(data_list[[i]]$metadata) 
}
if( nrow(unique(temp_data_list_qc)) != 1){stop("dimension mismatch between objects")}

data_list_concat <- data_list[[1]]
for(i in 2:length(data_list)){
  data_list_concat$TPM <- cbind(data_list_concat$TPM, data_list[[i]]$TPM)
  data_list_concat$metadata <- rbind(data_list_concat$metadata, data_list[[i]]$metadata)
}
data_list_concat$metadata <- as.data.frame(data_list_concat$metadata)

dataset_list$ENCODE_MM_DEV <- list(mat = data_list_concat$TPM, metadata=data_list_concat$metadata)
rm(data_list)
```


```{r, mouse adult single cell}
### USE THIS CODE AND LINKS TO DOWNLOAD AND PROCESSES EXPRESSION MATRIX IF NOT ALREADY SAVED
if(TRUE){
  ### if download speed via R is too slow, can also manually download files from https://figshare.com/articles/dataset/HCL_DGE_Data/7235471 ##
  download.file("https://figshare.com/ndownloader/files/22447898", paste(dirs$data_dir,"annotation_rmbatch_data_revised417.zip",sep="/"))
  unzip(paste(dirs$data_dir,"annotation_rmbatch_data_revised417.zip",sep="/"), exdir=dirs$data_dir)
  
  system(paste0("curl -L ",
                "https://figshare.com/ndownloader/files/23043329 ",
                "> data/dge_rmbatch_data.tar.gz"))
  untar(paste(data_dir,"dge_rmbatch_data.tar.gz",sep="/"),exdir=dirs$data_dir)
  
  files <- list.files(path=paste0(dirs$data_dir,"/annotation_rmbatch_data_revised417"),pattern="*.csv", full.names = T)
  for(f in files){
    temp <- fread(f)
    if(!exists("anno")){
      anno <- fread(f)
    } else{
      anno <- rbind(anno, temp,fill=TRUE)
    }
  }
  
  anno$V15 <- anno$V16 <- anno$V17 <- NULL
  anno$Celltype_comb <- tolower(str_replace(anno$Celltype, "_.*gh","" ))
  
  files <- list.files(path=paste0(dirs$data_dir,"/new"),pattern="*.txt.gz", full.names = T)
  rownames_genes <- foreach(f=files, .combine="c") %dopar% {
    rownames_genes <- fread(f)$V1
    rownames_genes
  }
  rownames_genes <- unique(rownames_genes)
  
  rm(exp_mat)
  ## foreach crashing r if called over too many files at once-  break into chunks and aggregate 1 chunk at a time
  file_chunks <- split(files, cut(seq_along(files), 10, labels = F))
  for(i in 1:length(file_chunks)){
    print(i)
    temp_mat <- foreach(f=file_chunks[[i]], .combine = "+") %dopar% {
        temp_mat <- matrix(0, nrow=length(rownames_genes), ncol=length(unique(anno$Celltype_comb)) )
        colnames(temp_mat) <- unique(anno$Celltype_comb)
        rownames(temp_mat) <- rownames_genes
        raw_exp <- fread(f)
        rownames(raw_exp) <- raw_exp$V1
        raw_exp$V1 <- NULL
        cts <- unique(anno$Celltype_comb[which(is.element(anno$Cell_id, colnames(raw_exp)))])
        for(ct in cts){
          temp_mat[rownames(raw_exp),ct] <-  temp_mat[rownames(raw_exp),ct]+
            rowSums(as.matrix(raw_exp)[,which(is.element(colnames(raw_exp),anno$Cell_id[which(anno$Celltype_comb==ct)]  ))])
        }
        temp_mat
    }
    if(!exists("exp_mat")){
      exp_mat <- temp_mat
    } else {
      exp_mat <- exp_mat + temp_mat
    }
  }
  
  
  mouse_exp_mat <- exp_mat
  #save(mouse_exp_mat, file = "mouse_sc.Rdata")
  
  #fwrite(mouse_exp_mat,file="mouse_sc.csv")
}
dataset_list$Adult_mouse_sc <- list(mat=exp_mat, meta=anno)
save(dataset_list, file = paste0(dirs$data_dir,"/mouse_transcriptome_data.Rdata"))
```
