---
title: "encoded_access_test"
output: html_document
---


```{r}
#minor change
#library(Seurat)
library(jsonlite)
library(httr)
library(stringr)
library(data.table)
library(doParallel)
library(biomaRt)
registerDoParallel(detectCores()-1)
```



```{r, create directory structure for project}
dirs <- list(data_dir = "./data",
             results_dir = "./results",
             figures_dir_parent = "./results/figures",
             figures_dir = "./results/figures/raw",
             tables_dir = "./results/tables")
dirs$data_dir <- "/u/scratch/l/leroybon/chrom_mod_proj_data"
for(d in dirs){
  if(!dir.exists(d)){dir.create(d)}
}

date <- format(Sys.time(),"%Y%m%d")

#load(paste0(dirs$data_dir,"/mouse_transcriptome_data.Rdata"))
```


```{r}
dataset_list <- list()
```

```{r, read early development}
## zygote, 2 cell, 4 cell, 8 cell, 16 cell, blasts cell

dirs$GSE45719 <- paste0(dirs$data_dir,"/GSE45719")
if(!(dir.exists(dirs$GSE45719))){
  dir.create(dirs$GSE45719)
  system(paste0("curl ",
                "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE45nnn/",
                "GSE45719/suppl/GSE45719_RAW.tar",
                paste0(" > ",dirs$GSE45719,"/GSE45719_RAW.tar" )))
  system(paste0("tar -xf ",dirs$GSE45719,"/GSE45719_RAW.tar -C ", dirs$GSE45719))
  
  files <- list.files(dirs$GSE45719)[grep(".txt.gz", list.files(dirs$GSE45719))]
  for(f in files){
   system(paste0("gunzip ", dirs$GSE45719,"/",f )) 
  }
}
temp_list <- list()
files <- list.files(dirs$GSE45719, full.names = T)[grep("expression.txt", list.files(dirs$GSE45719))]
temp_df <- str_split_fixed(str_split_fixed(files,"GSE45719/|_expression",3)[,2], "_|-", 5)[,1:4]
colnames(temp_df) <- c("GSM_ID","stage","rep","cell_id")
files_df <- data.frame(file=files, temp_df)
files_df <- files_df[!grepl("BXC|fibroblast|pooled|split|smartseq",files_df$file),]
for(i in 1:nrow(files_df)){
 temp_df <- fread(files_df$file[i])
 temp_df <- temp_df[,c("#Gene_symbol","reads")]; colnames(temp_df)[1] <- "gene_symbol"
 temp_df <- aggregate(reads ~ gene_symbol, temp_df,sum)
 temp_list[[paste(files_df[i,3:ncol(files_df)], collapse="_")]] <- temp_df
}
temp_df <- as.data.frame(matrix(nrow=length(temp_list[[1]]$gene_symbol), ncol=length(temp_list)))
rownames(temp_df) <- temp_list[[1]]$gene_symbol; colnames(temp_df) <- names(temp_list)
for(i in 1:length(temp_list)){
  if(any(temp_list[[i]]$gene_symbol != rownames(temp_df))){stop("gene names do not match")}
  temp_df[,i] <- temp_list[[i]]$reads
}

dataset_list$GSE45719 <- list(mat=temp_df,meta=NA)
```

```{r, read early gastrution paper ref data}
## e3.5, e4.5, e5.5, e6.5
target <- "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE100nnn/GSE100597/suppl/GSE100597%5Fcount%5Ftable%5FQC%5Ffiltered%2Etxt%2Egz"
temp <- tempfile()
download.file(target, temp)
temp_data <- read.table(temp, skip = 0, header = TRUE, sep = "\t")
rownames(temp_data) <- temp_data$X
temp_data$X <- NULL
temp_data[,1:ncol(temp_data)] <- apply(temp_data,2,as.numeric)

dataset_list[["GSE100597"]] <- list(mat=temp_data, meta=NA)
```

```{r, read early organogenesis paper ref data}
## Note this file is fairly large ~10 Gb
## e6.5, 7.25, e7.75, 8.5
library(data.table)
library(Seurat)

dirs$e_mtab_6967_atlas <- paste0(dirs$data_dir, "/e_mtab_6967_atlas")
if(!dir.exists(dirs$e_mtab_6967_atlas)){
  dir.create(dirs$e_mtab_6967_atlas)
  system(paste0("curl ",
                "https://www.ebi.ac.uk/arrayexpress/files/",
                "E-MTAB-6967/E-MTAB-6967.processed.1.zip ",
                "> ",dirs$e_mtab_6967_atlas ,"/atlas_data.zip"))
  system(paste0("unzip ",dirs$e_mtab_6967_atlas,"/atlas_data.zip -d ",dirs$e_mtab_6967_atlas))
  system(paste0("rm ",dirs$e_mtab_6967_atlas,"/atlas_data.zip"))
  system(paste0("tar -zxvf ",dirs$e_mtab_6967_atlas,"/atlas_data.tar.gz -C ",dirs$e_mtab_6967_atlas))
  system(paste0("mv ",dirs$e_mtab_6967_atlas,"/atlas",dirs$e_mtab_6967_atlas,"/e_mtab_6967_atlas"))
}

temp_data <- ReadMtx( paste0(dirs$e_mtab_6967_atlas,"/atlas/raw_counts.mtx"),
                     cells = paste0(dirs$e_mtab_6967_atlas,"/atlas/meta.csv"),
                     features = paste0(dirs$e_mtab_6967_atlas,"/atlas/genes.tsv"),
                     skip.cell = 1,
                     cell.sep = ",")

cell_meta <- fread(paste0(dirs$e_mtab_6967_atlas,"/atlas/meta.csv"))

which <- which(!cell_meta$doublet)
temp_data <- temp_data[,which]
cell_meta <- cell_meta[which,]

entities <- apply(cell_meta[,c("stage","cluster","cluster.sub","celltype","haem_subclust","endo_gutCluster")],1,paste,collapse=",")
tempdf <- unique(cell_meta[,c("stage","cluster","cluster.sub","celltype","haem_subclust","endo_gutCluster")])
entity_types <- unique(entities)
cell_entity_map <- list()
for(i in 1:length(entity_types)){
  which <- which(entities == entity_types[i])
  cell_entity_map[[entity_types[i]]] <- cell_meta$cell[which]
}

df <- data.frame(matrix(nrow = nrow(temp_data), ncol=length(entity_types)))
rownames(df) <- rownames(temp_data)
colnames(df) <- names(cell_entity_map)

for(i in 1:ncol(df)){
  if(i %% 50 ==0){print(i)}
  df[,i] <- rowSums(temp_data[,cell_entity_map[[i]],drop=F])
}

for(i in 1:ncol(df)){
  tempdf$num_cells[i] <- length(cell_entity_map[[i]])
}

dataset_list$E_MTAB_6967 <- list(mat=df,meta=tempdf) 
rm(temp_data)
```



```{r, read mouse organogenesis paper}
# E9.5 to E13.5

dirs$GSE119945 <- paste0(dirs$data_dir,"/GSE119945/")
if(!(dir.exists(dirs$GSE119945))){
  dir.create(dirs$GSE119945)
  system(paste0("curl ",
                "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE119nnn/",
                "GSE119945/suppl/GSE119945%5Fgene%5Fcount%2Etxt%2Egz",
                paste0("> ",dirs$GSE119945,"gene_count.txt.gz" )))
  system(paste0("gunzip -c ",dirs$GSE119945,"gene_count.txt.gz > ",dirs$GSE119945,"gene_count.txt"))
  system(paste0("curl ",
                "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE119nnn/",
                "GSE119945/suppl/GSE119945%5Fcell%5Fannotate%2Ecsv%2Egz",
                paste0("> ",dirs$GSE119945,"cell_annotate.csv.gz" )))
  system(paste0("curl ",
                "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE119nnn/",
                "GSE119945/suppl/GSE119945%5Fgene%5Fannotate%2Ecsv%2Egz",
                paste0("> ",dirs$GSE119945,"gene_annotate.csv.gz" )))
}

temp_cells <- fread(paste0(dirs$GSE119945,"cell_annotate.csv.gz" ))
temp_cells <- as.data.frame(temp_cells)
temp_cells$Sub_Cluster[grep("unknown", temp_cells$Sub_Cluster)] <- "0"
temp_features <- fread(paste0(dirs$GSE119945,"gene_annotate.csv.gz" ))
temp_features <- as.data.frame(temp_features)
temp_cells$cluster <- str_remove_all(apply(temp_cells[,c("day","Main_Cluster","Sub_Cluster")],
                            1,paste,collapse="-")," ")
temp_cluster_exp_mat <- matrix(0,nrow = nrow(temp_features),
                          ncol = length(unique(temp_cells$cluster)) )
rownames(temp_cluster_exp_mat) <- temp_features$gene_id
colnames(temp_cluster_exp_mat) <- unique(temp_cells$cluster)
## num lines gene_count.txt.gz is at max cells x features + 2
temp_nrow <- 500e6 
temp_num <- ceiling(file.size(paste0(dirs$GSE119945,"gene_count.txt")) / temp_nrow)

dirs$temp_split <- paste0(dirs$GSE119945,"temp_split")
if(!dir.exists(dirs$temp_split)){
  dir.create(dirs$temp_split)
  system(paste0("split ",dirs$GSE119945,"gene_count.txt",
                " -n ", "l/",temp_num, " ",
                paste0(dirs$temp_split,"/chunk_")
                )
  )
}


file_list <- list.files(dirs$temp_split, full.names = T) 
file_list <- file_list[grep("chunk_", file_list)]

dataset_list$GSE119945$mat <- foreach( f=file_list, .combine = "+") %dopar% {
  if(grepl("chunk_aa",f)){
     # temp_header <- system(paste0("head -n 2 ", f), intern = T)
     # fileConn <- file(paste0(dirs$GSE119945,"temp_split/","header.txt"))
     # writeLines(temp_header, fileConn)
     # close(fileConn)
     mat_part <- fread(f, skip = 2, header = F,
                    col.names = c("gene","cell","value") )
   } else {
     mat_part <- fread(f, skip = 0, header = F,
                    col.names = c("gene","cell","value") )
   }
  mat_part <- as.data.frame(mat_part)
  mat_part$cluster <- temp_cells$cluster[match(mat_part$cell, 1:nrow(temp_cells))]
  ## next need to aggregate counts for each gene for each cluster ....
  
  
  # temp_mp<- aggregate(value ~ gene + cluster, data = mat_part, FUN=sum)
  
  temp_clusters <- unique(mat_part$cluster)
  for(i in 1:length(temp_clusters)){
    mat_part_sub <- mat_part[which(mat_part$cluster==temp_clusters[i]),]
    temp_mp <- aggregate(value ~ gene + cluster, data = mat_part_sub, FUN=sum)
    temp_cluster_exp_mat[temp_mp$gene, unique(temp_mp$cluster)] <-
      temp_cluster_exp_mat[temp_mp$gene, unique(temp_mp$cluster)] + temp_mp$value
  }
  temp_cluster_exp_mat
}

dataset_list$GSE119945$meta <- temp_cells
rm(temp_cells)
```


```{r, read encode mouse ref RNAseq}

## e10.5, e11.5, e12.5, e13.5, e14.5, e15.5, e16.5, neonat, 8wk adult, 24wk adult
## get ENCODE search results : mouse epigenome reference RNAseq
search_query <- "https://www.encodeproject.org/search/?type=Experiment&related_series.@type=ReferenceEpigenome&status=released&replicates.library.biosample.donor.organism.scientific_name=Mus+musculus&assay_title=polyA+plus+RNA-seq&frame=object&format=json&limit=all"
query_result <- GET(search_query)
query_result <- content(query_result , "text", encoding = "UTF-8")
query_result <- fromJSON(query_result, flatten = T)
query_table <- query_result$`@graph`[!unlist(lapply(query_result$`@graph`, is.list))]

##
data_list <- list()
for(i in 1:nrow(query_table)){
  if(i %% 5 == 0){print(i)}
  target <- paste0("https://www.encodeproject.org/",
                          query_table$accession[i],
                          "/?frame=embedded&format=json")
  target <- GET(target)
  target_result <- content(target, "text", encoding = "UTF-8")
  target_result <- fromJSON(target_result, flatten = TRUE)
  temp_gene_anno <-  unique(target_result$files$genome_annotation)
  temp_gene_anno <- temp_gene_anno[grep("ENSEMBL", temp_gene_anno,invert = T)]
  which <- which(target_result$files$file_format=="tsv" &
                 target_result$files$output_type=="gene quantifications" &
                 grepl(max(as.numeric(str_remove_all(temp_gene_anno,"[a-zA-Z]")),na.rm = T),
                       target_result$files$genome_annotation) &
                 target_result$files$status == "released")
  which_cols <- c("accession", "href","biosample_ontology",
                "biological_replicates", "technical_replicates",
                "biological_replicates_formatted", "donors",
                "dataset")
  gene_quant_files <- target_result$files[which,which_cols]
  gene_quant_files[,which(unlist(lapply(gene_quant_files, is.list)))] <- 
    sapply(gene_quant_files[,which(unlist(lapply(gene_quant_files, is.list)))], paste )
  target_metadata <- cbind(query_table[i,], gene_quant_files)
  
  for(j in 1:nrow(target_metadata)){
    temp <- tempfile()
    download.file(paste0("https://www.encodeproject.org/",target_metadata$href[j]), temp)
    temp_data <- read.table(temp, skip = 0, header = TRUE, sep = "\t")
    data_list[[length(data_list)+1]] <- list(TPM = data.frame(TPM=temp_data$TPM),
                                             metadata = cbind(target_metadata[j,], read_count=sum(temp_data$posterior_mean_count) ) )
    rownames(data_list[[length(data_list)]]$TPM) <- temp_data$gene_id
  }
}

temp_data_list_qc <- data.frame(tpm_nrow=numeric(length=length(data_list)),
                                tpm_ncol=numeric(length=length(data_list)),
                                metadata_nrow=numeric(length=length(data_list)))
for(i in 1:length(data_list)){
  temp_data_list_qc$tpm_nrow[i] <- nrow(data_list[[i]]$TPM)
  temp_data_list_qc$tpm_ncol[i] <- ncol(data_list[[i]]$TPM)
  temp_data_list_qc$metadata_nrow[i] <- nrow(data_list[[i]]$metadata) 
}
if( nrow(unique(temp_data_list_qc)) != 1){stop("dimension mismatch between objects")}

data_list_concat <- data_list[[1]]
for(i in 2:length(data_list)){
  data_list_concat$TPM <- cbind(data_list_concat$TPM, data_list[[i]]$TPM)
  data_list_concat$metadata <- rbind(data_list_concat$metadata, data_list[[i]]$metadata)
}
data_list_concat$metadata <- as.data.frame(data_list_concat$metadata)

dataset_list$ENCODE_MM_DEV <- list(mat = data_list_concat$TPM, metadata=data_list_concat$metadata)
rm(data_list)
```


```{r, mouse adult single cell}
### USE THIS CODE AND LINKS TO DOWNLOAD AND PROCESSES EXPRESSION MATRIX IF NOT ALREADY SAVED
dirs$GSE108097 <- paste0(dirs$data_dir,"/GSE108097/")
if(!(dir.exists(dirs$GSE108097))){
  dir.create(dirs$GSE108097)
}
if(TRUE){
  ### if download speed via R is too slow, can also manually download files from https://figshare.com/articles/dataset/HCL_DGE_Data/7235471 ##
  download.file("https://figshare.com/ndownloader/files/10756795",paste(dirs$GSE108097,"MCA_BatchRemove_dge.zip",sep="/"))
  download.file("https://figshare.com/ndownloader/files/11083451", paste(dirs$GSE108097,"MCA_CellAssignments.csv",sep="/"))
  unzip(paste(dirs$GSE108097,"MCA_BatchRemove_dge.zip",sep="/"), exdir=dirs$GSE108097)
  
  
  anno <- fread(paste(dirs$GSE108097,"MCA_CellAssignments.csv",sep="/"),header=T)
  
  files <- list.files(path=paste0(dirs$GSE108097,"/rmbatch_dge"),pattern="*.txt.gz", full.names = T)
  rownames_genes <- foreach(f=files, .combine="c") %dopar% {
    rownames_genes <- fread(f)$V1
    rownames_genes
  }
  rownames_genes <- unique(rownames_genes)
  
  rm(exp_mat)
  ## foreach crashing r if called over too many files at once-  break into chunks and aggregate 1 chunk at a time
  file_chunks <- split(files, cut(seq_along(files), 10, labels = F))
  for(i in 1:length(file_chunks)){
    print(i)
    temp_mat <- foreach(f=file_chunks[[i]], .combine = "+") %dopar% {
        temp_mat <- matrix(0, nrow=length(rownames_genes), ncol=length(unique(anno$Annotation)) )
        colnames(temp_mat) <- unique(anno$Annotation)
        rownames(temp_mat) <- rownames_genes
        raw_exp <- as.data.frame(fread(f))
        rownames(raw_exp) <- raw_exp$V1
        raw_exp$V1 <- NULL
        cts <- unique(anno$Annotation[which(is.element(anno$Cell.name, colnames(raw_exp)))])
        for(ct in cts){
          temp_mat[rownames(raw_exp),ct] <-  temp_mat[rownames(raw_exp),ct]+
            rowSums(as.matrix(raw_exp)[,which(is.element(colnames(raw_exp),anno$Cell.name[which(anno$Celltype_comb==ct)]  ))])
        }
        temp_mat
    }
    if(!exists("exp_mat")){
      exp_mat <- temp_mat
    } else {
      exp_mat <- exp_mat + temp_mat
    }
  }
  mouse_exp_mat <- exp_mat
  #save(mouse_exp_mat, file = "mouse_sc.Rdata")
  
  #fwrite(mouse_exp_mat,file="mouse_sc.csv")
}
dataset_list$GSE108097 <- list(mat=exp_mat, meta=anno)
save(dataset_list, file = paste0(dirs$data_dir,"/mouse_transcriptome_data.Rdata"))


```


```{r, temp cleaning developmental_time chunk}
#"a.01.zygote"     "a.02.2cell"      "a.03.4cell"      "a.04.8cell"      "a.05.16cell"     "a.06.morula"     "a.07.blastocyst"
### std_dev_time : a01.zygote ... (stage named by early morphology), e04.50 (embronic day.day_frac), p0.00.0 (postnatal year.week.day) 
### std_dev_time_modifier : early, mid (default), late


## organize data here but move to original chunks later
temp_ds <- dataset_list$E_MTAB_6967
head(temp_ds$mat)
head(temp_ds$meta)
col_data <- data.frame(colnames=colnames(temp_ds$mat))
unique(temp_ds$meta$stage)

temp_df <- data.frame(std_dev_times = c("unknown","e06.50","e06.75","e07.00","e07.25","e07.50","e07.75","e08.00","e08.25","e08.50"),
                      provided_dev_times = c("mixed_gastrulation","E6.5","E6.75","E7.0","E7.25","E7.5","E7.75","E8.0","E8.25","E8.5"))
head(col_data)
col_data$dev_time_temp <- temp_ds$meta$stage
col_data$std_dev_time <- temp_df$std_dev_times[match(col_data$dev_time_temp,temp_df$provided_dev_times)]
col_data$dev_time_temp <- NULL
dataset_list$E_MTAB_6967$col_data <- col_data


temp_ds <- dataset_list$GSE100597
## NOTE need to use markers from study to classify lineages in matrix for clustering 
head(temp_ds$mat)
head(temp_ds$meta)
col_data <- data.frame(colnames=colnames(temp_ds$mat))
col_data$dev_time_temp <- str_split_fixed(colnames(temp_ds$mat),"_",2)[,1]
unique(col_data$dev_time_temp)

temp_df <- data.frame(std_dev_times = c("e03.50","e04.50","e05.50","e06.50","e06.75"),
                      provided_dev_times = c("E3.5","E4.5","E5.5","E6.5","E6.75"))

col_data$std_dev_time <- temp_df$std_dev_times[match(col_data$dev_time_temp,temp_df$provided_dev_times)]
col_data$dev_time_temp <- NULL
dataset_list$GSE100597$col_data <- col_data





temp_ds <- dataset_list$GSE119945
head(temp_ds$mat)
head(temp_ds$meta)
## need to go back and set columns to be day-cluster_major-cluster_minor
col_data <- data.frame(colnames=colnames(temp_ds$mat))
col_data$dev_time_temp <- str_split_fixed(colnames(temp_ds$mat),"-",2)[,1]
sort(unique(col_data$dev_time_temp))

temp_df <- data.frame(std_dev_times = c("e09.50","e10.50","e11.50","e12.50","e13.50"),
                      provided_dev_times = c("9.5","10.5","11.5","12.5","13.5"))

col_data$std_dev_time <- temp_df$std_dev_times[match(col_data$dev_time_temp,temp_df$provided_dev_times)]
col_data$dev_time_temp <- NULL
dataset_list$GSE119945$col_data <- col_data



temp_ds <- dataset_list$GSE108097
head(temp_ds$mat)
head(temp_ds$meta)
## need to go back and set columns to be day-cluster_major-cluster_minor
col_data <- data.frame(colnames=colnames(temp_ds$mat))
col_data$dev_time_temp <- tolower(str_extract(col_data$colnames,"[Nn]eonat|[Ff][o]*etal"))
col_data$dev_time_temp[which(is.na(col_data$dev_time_temp))] <-"adult"
sort(unique(col_data$dev_time_temp))

temp_df <- data.frame(std_dev_times = c("e14.50","p0.00.00","p0.08.00"),
                      provided_dev_times = c("fetal","neonat","adult"))

col_data$std_dev_time <- temp_df$std_dev_times[match(col_data$dev_time_temp,temp_df$provided_dev_times)]
col_data$dev_time_temp <- NULL
dataset_list$GSE108097$col_data <- col_data



temp_ds <- dataset_list$GSE45719
head(temp_ds$mat)
head(temp_ds$meta)
## need to go back and set columns to be day-cluster_major-cluster_minor
col_data <- data.frame(colnames=colnames(temp_ds$mat))
col_data$dev_time_temp <- str_split_fixed(col_data$colnames,"_",2)[,1]
col_data$dev_time_temp[grep("zy",col_data$dev_time_temp)] <- "zy"
sort(unique(col_data$dev_time_temp))

#"a01.zygote"     "a02.2cell"      "a03.4cell"      "a04.8cell"      "a05.16cell"     "a06.morula"     "a07.blastocyst"
temp_df <- data.frame(std_dev_times = c("a01.zygote","a02.2cell.1.early","a02.2cell.2.mid","a02.2cell.3.late", "unknown",
                                        "a03.4cell","a04.8cell","a05.16cell",
                                        "a07.blastocyst.1.early","a07.blastocyst.2.mid","a07.blastocyst.3.late" ),
                      provided_dev_times = c("zy","early2cell","mid2cell","late2cell","C57twocell",
                                        "4cell","8cell","16cell",
                                        "earlyblast","midblast","lateblast"))

col_data$std_dev_time <- temp_df$std_dev_times[match(col_data$dev_time_temp,temp_df$provided_dev_times)]
col_data$dev_time_temp <- NULL
dataset_list$GSE45719$col_data <- col_data




dataset_list$ENCODE_MM_DEV$metadata$col_names <- apply(
  dataset_list$ENCODE_MM_DEV$metadata[,c("accession","biological_replicates")],
  1,paste,collapse="_")
colnames(dataset_list$ENCODE_MM_DEV$mat) <- dataset_list$ENCODE_MM_DEV$metadata$col_names
temp_ds <- dataset_list$ENCODE_MM_DEV
head(temp_ds$mat)
head(temp_ds$meta)
## need to go back and set columns to be day-cluster_major-cluster_minor

col_data <- data.frame(colnames=colnames(temp_ds$mat))
col_data$dev_time_temp <- temp_ds$metadata$life_stage_age
sort(unique(col_data$dev_time_temp))

#"a01.zygote"     "a02.2cell"      "a03.4cell"      "a04.8cell"      "a05.16cell"     "a06.morula"     "a07.blastocyst"
temp_df <- data.frame(std_dev_times = c("e11.50","e13.50","e14.50","e15.50", "e16.50",
                                        "p0.00.00","p0.08.00","p0.24.00"),
                      provided_dev_times = c("embryonic 11.5 days","embryonic 13.5 days","embryonic 14.5 days","embryonic 15.5 days",
                                             "embryonic 16.5 days","postnatal 0 days", "adult 8 weeks","adult 24 weeks"))

col_data$std_dev_time <- temp_df$std_dev_times[match(col_data$dev_time_temp,temp_df$provided_dev_times)]
col_data$dev_time_temp <- NULL
dataset_list$ENCODE_MM_DEV$col_data <- col_data




```


```{r,temp cleaning gene_annotations chunk}
## organize data here but move to original chunks later

## gene_df <- get from Biomart 

## import ensembl gene data
ensembl = useEnsembl(biomart="ensembl", dataset="mmusculus_gene_ensembl")
genes <- getBM(attributes=c('chromosome_name','start_position','end_position',
                            'external_gene_name', 'ensembl_gene_id','gene_biotype',
                            'external_synonym'),
                 #filters = list('biotype'='protein_coding'),
                 mart = ensembl, useCache = F) 
genes <- genes[which(is.element(genes$chromosome_name, c(1:22, "X", "Y", "MT")) & genes$external_gene_name != "" ) ,]
temp_genes <- as.data.frame(unique(rbind(as.matrix(genes[,c("ensembl_gene_id","external_gene_name")]),
                                         as.matrix(genes[,c("ensembl_gene_id","external_synonym")]))))


### check how many gene names can be match between Biomart and matrix
temp <- make.unique(temp_genes$ensembl_gene_id[match(rownames(dataset_list$E_MTAB_6967$mat), temp_genes$ensembl_gene_id)])
temp[which(is.na(temp))] <- "NA"
rownames(dataset_list$E_MTAB_6967$mat) <- temp


rownames(dataset_list$GSE100597$mat)[sample(1:1e4, 25)]
temp <- str_split_fixed(rownames(dataset_list$GSE100597$mat),"_",2)[,1]
temp <- make.unique(temp_genes$ensembl_gene_id[match(temp, temp_genes$ensembl_gene_id)])
temp[which(is.na(temp))] <- "NA" 
rownames(dataset_list$GSE100597$mat) <- temp


temp <- str_split_fixed(rownames(dataset_list$GSE119945$mat),"\\.",2)[,1]
temp <- make.unique(temp_genes$ensembl_gene_id[match(temp, temp_genes$ensembl_gene_id)])
temp[which(is.na(temp))] <- "NA"
rownames(dataset_list$GSE119945$mat) <- temp


temp <- str_split_fixed(rownames(dataset_list$ENCODE_MM_DEV$mat),"\\.",2)[,1]
temp <- make.unique(temp_genes$ensembl_gene_id[match(temp, temp_genes$ensembl_gene_id)])
temp[which(is.na(temp))] <- "NA"
rownames(dataset_list$ENCODE_MM_DEV$mat) <- temp


temp <- rownames(dataset_list$GSE45719$mat)
temp <- make.unique(temp_genes$ensembl_gene_id[match(temp, temp_genes$ensembl_gene_id)])
temp[which(is.na(temp))] <- "NA" 
rownames(dataset_list$GSE45719$mat) <- temp


temp <- rownames(dataset_list$GSE108097$mat)
temp <- make.unique(temp_genes$ensembl_gene_id[match(temp, temp_genes$ensembl_gene_id)])
temp[which(is.na(temp))] <- "NA" 
rownames(dataset_list$GSE108097$mat) <- temp 


```



```{r, make colnames cleaner - add to individual data read ins later}

sample(colnames(dataset_list$E_MTAB_6967$mat),10)
head(dataset_list$E_MTAB_6967$col_data)
dataset_list$E_MTAB_6967$col_data$informative_colname <- apply(str_split_fixed(dataset_list$E_MTAB_6967$col_data$colnames,",",6)[,c(1,4,2,3)],
                                                               1,paste,collapse="_",sep="")

sample(colnames(dataset_list$GSE100597$mat),10) ## current names uninformative


temp_file <- tempfile()
download.file("https://ars.els-cdn.com/content/image/1-s2.0-S2211124717309610-mmc2.xlsx",temp_file)
temp_marker_genes <- as.data.frame(readxl::read_excel(temp_file,skip=1,
                                                      col_types = c("text","text","numeric","numeric",rep("skip",7))))
temp_marker_genes$gene_ensembl_id <- temp_genes$ensembl_gene_id[match(temp_marker_genes$Gene, temp_genes$external_gene_name)]
unlink(temp_file)
cluster_names <- unique(temp_marker_genes$Stage)

dim(dataset_list$GSE100597$mat)

temp_marker_genes$gene_ensembl_id <- temp_genes$ensembl_gene_id[match(temp_marker_genes$Gene, temp_genes$external_gene_name)]
temp_cell_classifier <- cbind(dataset_list$GSE100597$col_data,
                              matrix(nrow = nrow(dataset_list$GSE100597$col_data),
                                     ncol = length(cluster_names),dimnames = list(NULL,cluster_names)) )
temp_marker_genes_list <- list()
for(i in 1:length(cluster_names)){
  temp_marker_genes_list[[ cluster_names[i] ]] <- temp_marker_genes[which(temp_marker_genes$Stage==cluster_names[i]),]
}
for(i in 1:ncol(dataset_list$GSE100597$mat)){
  if(i %% 30 == 1){print(i)}
  temp_cell_vec <- dataset_list$GSE100597$mat[,i,drop=F]
  for(j in 1:length(cluster_names)){
    colnames(temp_cell_classifier)
    temp <- temp_cell_vec[temp_marker_genes_list[[cluster_names[j]] ]$gene_ensembl_id,1]
    temp_cell_classifier[i, cluster_names[j]] <- sum(log10(temp[!is.na(temp)]+1))/length(temp)
  }
}


sample(colnames(dataset_list$GSE119945$mat),10) ## current names uninformative

sample(colnames(dataset_list$ENCODE_MM_DEV$mat),10) ## current names uninformative

sample(colnames(dataset_list$GSE45719$mat),10)

sample(colnames(dataset_list$GSE108097$mat),10)
```


```{r, }


## subset matrices for only overlapping genes so that rows match


## concatenate all matrices and col_data for name and dev_time

## gene_specificity functions to get weights --> sim tree (virtual parent nodes)

## bipartite cluster prev to next stage  --> dev tree (real parent nodes)

## 

##
```



```{r, standardize column / cell/ tissue names via uberon }

## download uberon file
dirs$uberon_dir <- paste0(dirs$data_dir,"/uberon")
if(!dir.exists(dirs$uberon_dir)){
  dir.create(dirs$uberon_dir)
}

# system(paste0("curl -L ",
#               "https://github.com/obophenotype/uberon/releases/latest/download/uberon.owl",
#               paste0(" > ",dirs$uberon_dir,"/uberon.owl" )))


library(ontologyIndex)
onto <- get_ontology("https://github.com/obophenotype/uberon/releases/latest/download/uberon.obo")

# dataset_list$E_MTAB_6967$meta$celltype  ## mappable
# dataset_list$GSE100597$col_data  ## no cell type labels
# dataset_list$GSE119945$meta ## no cell type labels
# dataset_list$ENCODE_MM_DEV$metadata$biosample_ontology  ## UBERON_IDs
# dataset_list$GSE45719$col_data  ## mappable
# dataset_list$GSE108097$col_data ## mappable 

### CellMarker database to label cluster/ check whether cell types match - also check into papers and how types were assigned 

```

