---
title: "encoded_access_test"
output: html_document
---


```{r}
library(jsonlite)
library(httr)
library(stringr)
```

```{r, read early development}
## zygote, 2 cell, 4 cell, 8 cell, 16 cell, 32 cell

```

```{r, read early gastrution paper ref data}
## e3.5, e4.5, e5.5, e6.5
target <- "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE100nnn/GSE100597/suppl/GSE100597%5Fcount%5Ftable%5FQC%5Ffiltered%2Etxt%2Egz"
temp <- tempfile()
download.file(target, temp)
temp_data <- read.table(temp, skip = 0, header = TRUE, sep = "\t")

```

```{r, read organogenesis paper ref data}
## e6.5, 7.25, e7.75, 8.5
library(data.table)
library(Seurat)
target <-"C:/Users/LeroyBondhus/Downloads/E-MTAB-6967.processed.1/atlas/raw_counts.mtx"
cells <-"C:/Users/LeroyBondhus/Downloads/E-MTAB-6967.processed.1/atlas/meta.csv"
features <- "C:/Users/LeroyBondhus/Downloads/E-MTAB-6967.processed.1/atlas/genes.tsv"
temp <- tempfile()
download.file(target, temp)
temp_data <- ReadMtx(target,cells = cells, features = features)
```

```{r, read encode mouse ref RNAseq}

## e10.5, e11.5, e12.5, e13.5, e14.5, e15.5, e16.5, neonat, 8wk adult, 24wk adult
## get ENCODE search results : mouse epigenome reference RNAseq
search_query <- "https://www.encodeproject.org/search/?type=Experiment&related_series.@type=ReferenceEpigenome&status=released&replicates.library.biosample.donor.organism.scientific_name=Mus+musculus&assay_title=polyA+plus+RNA-seq&frame=object&format=json&limit=all"
query_result <- GET(search_query)
query_result <- content(query_result , "text", encoding = "UTF-8")
query_result <- fromJSON(query_result, flatten = T)
query_table <- query_result$`@graph`[!unlist(lapply(query_result$`@graph`, is.list))]

##
data_list <- list()
for(i in 1:nrow(query_table)){
  if(i %% 5 == 0){print(i)}
  target <- paste0("https://www.encodeproject.org/",
                          query_table$accession[i],
                          "/?frame=embedded&format=json")
  target <- GET(target)
  target_result <- content(target, "text", encoding = "UTF-8")
  target_result <- fromJSON(target_result, flatten = TRUE)
  temp_gene_anno <-  unique(target_result$files$genome_annotation)
  temp_gene_anno <- temp_gene_anno[grep("ENSEMBL", temp_gene_anno,invert = T)]
  which <- which(target_result$files$file_format=="tsv" &
                 target_result$files$output_type=="gene quantifications" &
                 grepl(max(as.numeric(str_remove_all(temp_gene_anno,"[a-zA-Z]")),na.rm = T),
                       target_result$files$genome_annotation) &
                 target_result$files$status == "released")
  which_cols <- c("accession", "href","biosample_ontology",
                "biological_replicates", "technical_replicates",
                "biological_replicates_formatted", "donors",
                "dataset")
  gene_quant_files <- target_result$files[which,which_cols]
  gene_quant_files[,which(unlist(lapply(gene_quant_files, is.list)))] <- 
    sapply(gene_quant_files[,which(unlist(lapply(gene_quant_files, is.list)))], paste )
  target_metadata <- cbind(query_table[i,], gene_quant_files)
  
  for(j in 1:nrow(target_metadata)){
    temp <- tempfile()
    download.file(paste0("https://www.encodeproject.org/",target_metadata$href[j]), temp)
    temp_data <- read.table(temp, skip = 0, header = TRUE, sep = "\t")
    data_list[[length(data_list)+1]] <- list(TPM = data.frame(TPM=temp_data$TPM),
                                             metadata = target_metadata[j,])
    rownames(data_list[[length(data_list)]]$TPM) <- temp_data$gene_id
  }
}

temp_data_list_qc <- data.frame(tpm_nrow=numeric(length=length(data_list)),
                                tpm_ncol=numeric(length=length(data_list)),
                                metadata_nrow=numeric(length=length(data_list)))
for(i in 1:length(data_list)){
  temp_data_list_qc$tpm_nrow[i] <- nrow(data_list[[i]]$TPM)
  temp_data_list_qc$tpm_ncol[i] <- ncol(data_list[[i]]$TPM)
  temp_data_list_qc$metadata_nrow[i] <- nrow(data_list[[i]]$metadata) 
}
if( nrow(unique(temp_data_list_qc)) != 1){stop("dimension mismatch between objects")}

data_list_concat <- data_list[[1]]
for(i in 2:length(data_list)){
  data_list_concat$TPM <- cbind(data_list_concat$TPM, data_list[[i]]$TPM)
  data_list_concat$metadata <- rbind(data_list_concat$metadata, data_list[[i]]$metadata)
}
data_list_concat$metadata <- as.data.frame(data_list_concat$metadata)
#rm(data_list)

```



```{r, }


```