---
title: "general_summary_analyses"
author: "Leroy Bondhus"
date: "2023-07-10"
output: html_document
---

```{r, gene length barcode plot, pretty}
library(cowplot)

#### add to density plot line for all genes

df <- cbind(gene_sets$genes_all[0,],
            data.frame(gene_length = numeric(), chrom_subset_name = character()))
for(i in 1:length(gene_sets$human_chromatin_modifiers$gene_subsets)){
  ## get current subset of genes
  chrom_subset <- gene_sets$human_chromatin_modifiers$gene_subsets[[i]] 
  chrom_subset_name <- names(gene_sets$human_chromatin_modifiers$gene_subsets)[i]
  if(length(chrom_subset)==0){next}
  ## map current subset to table with gene lenght info
  temp_df <- gene_sets$genes_all[which(is.element(gene_sets$genes_all$ensembl_gene_id, chrom_subset )),]
  temp_df$gene_length <- temp_df$end_position - temp_df$start_position
  temp_df$chrom_subset_name <- chrom_subset_name
  df <- rbind(df, temp_df)
}

ggl <- list()
ggl[["all_density"]] <- ggplot(df[grep("^all_",df$chrom_subset_name),], aes(x=log10(gene_length)))+
  coord_cartesian(xlim = c(2, 6.5))+
  geom_density(alpha=0.5)+
  barcode_theme+
  theme(axis.title.x = element_blank())

ggl[["barcodes"]] <- ggplot(df, aes(x=log10(gene_length),xend=log10(gene_length), y=0,yend=1))+
  geom_segment(alpha=0.5)+
  coord_cartesian(ylim = c(0.2,0.5), xlim = c(2, 6.5))+
  barcode_theme+
  facet_grid(chrom_subset_name ~ . )+
  theme(strip.text.y = element_text(angle = 0))
  
plot_grid(ggl$all_density, ggl$barcodes, align="v",axis="r",nrow = 2, rel_heights = c(2,8))

```

```{r, transcript length barcode plot, pretty}
library(cowplot)

#### add to density plot line for all genes

df <- data.frame(ensembl_gene_id = character(),
                 transcript_length = numeric(),
                 chrom_subset_name = character())
for(i in 1:length(gene_sets$human_chromatin_modifiers$gene_subsets)){
  ## get current subset of genes
  chrom_subset <- gene_sets$human_chromatin_modifiers$gene_subsets[[i]] 
  chrom_subset_name <- names(gene_sets$human_chromatin_modifiers$gene_subsets)[i]
  if(length(chrom_subset)==0){next}
  ## map current subset to table with transcript length info
  temp_df <- gene_sets$genes_all_transcript[which(is.element(gene_sets$genes_all_transcript$ensembl_gene_id, chrom_subset )),]
  temp_df <- unique(temp_df[,c("ensembl_gene_id", "transcript_length")])
  temp_df$chrom_subset_name <- chrom_subset_name
  df <- rbind(df, temp_df)
}

ggl <- list()
ggl[["all_density"]] <- ggplot(df[grep("^all_",df$chrom_subset_name),], aes(x=log10(transcript_length)))+
  coord_cartesian(xlim = c(2, 6.5))+
  geom_density(alpha=0.5)+
  barcode_theme+
  theme(axis.title.x = element_blank())

ggl[["barcodes"]] <- ggplot(df, aes(x=log10(transcript_length),xend=log10(transcript_length), y=0,yend=1))+
  geom_segment(alpha=0.5)+
  coord_cartesian(ylim = c(0.2,0.5), xlim = c(2, 6.5))+
  barcode_theme+
  facet_grid(chrom_subset_name ~ . )+
  theme(strip.text.y = element_text(angle = 0))
  
plot_grid(ggl$all_density, ggl$barcodes, align="v",axis="r",nrow = 2, rel_heights = c(2,8))

```

```{r, first and last exon v all exons, in progress}

df <- data.frame(ensembl_gene_id = character(),
                 exon_length = character(),
                 exon_count = numeric(),
                 rank = numeric(),
                 rank_from_end = numeric(),
                 chrom_subset_name = character())
for(i in 1:length(gene_sets$human_chromatin_modifiers$gene_subsets)){
  ## get current subset of genes
  chrom_subset <- gene_sets$human_chromatin_modifiers$gene_subsets[[i]] 
  chrom_subset_name <- names(gene_sets$human_chromatin_modifiers$gene_subsets)[i]
  if(length(chrom_subset)==0){next}
  ## map current subset to table with exon info
  
  ## exon length
  temp_df <- gene_sets$genes_all_transcript[which(is.element(gene_sets$genes_all_transcript$ensembl_gene_id,
                                                        chrom_subset )),]
  temp_df$exon_length <- temp_df$exon_chrom_end - temp_df$exon_chrom_start
  temp_df$gene_factor_num <- as.numeric(as.factor(temp_df$ensembl_gene_id))
  temp_name <- paste0(chrom_subset_name, ": ", "exon_length")
  
  ## exon count
  temp_df_2 <- as.data.frame(table(temp_df$ensembl_transcript_id))
  colnames(temp_df_2)[1] <- "ensembl_transcript_id"
  colnames(temp_df_2)[2] <- "exon_count"
  temp_df <- merge(temp_df, temp_df_2, sort=FALSE)
  
  ## exon rank from end
  temp_df$rank_from_end <- temp_df$exon_count - temp_df$rank + 1
  
  temp_df$chrom_subset_name <- chrom_subset_name
  df <- rbind(df, temp_df[,which(is.element(colnames(temp_df),colnames(df)))])
}

ggl <- list()
df$is_last <- df$rank_from_end == 1 & df$rank != 1
df$is_first <- df$rank == 1 & df$rank_from_end != 1
ggplot(df[grep("^all_",df$chrom_subset_name),], aes(x=log10(exon_length), color=is_last))+
  coord_cartesian(xlim = c(0, 5))+
  geom_density(alpha=0.5)+
  barcode_theme+
  theme(axis.title.x = element_blank())

ggplot(df[grep("^all_",df$chrom_subset_name),], aes(x=log10(exon_length), color=is_first))+
  coord_cartesian(xlim = c(0, 5))+
  geom_density(alpha=0.5)+
  barcode_theme+
  theme(axis.title.x = element_blank())


```



```{r, look at basic chromatin modifier summary statistics}

## gene length
## transcript length
## exon count
## exon size
ggl <- list()
for(i in 14){ #  1:length(gene_sets$human_chromatin_modifiers$gene_subsets)){
  chrom_subset <- gene_sets$human_chromatin_modifiers$gene_subsets[[i]] 
  chrom_subset_name <- names(gene_sets$human_chromatin_modifiers$gene_subsets)[i]
  if(length(chrom_subset)==0){next}
  ## gene length density + barcode plots
  df <- gene_sets$genes_all[which(is.element(gene_sets$genes_all$ensembl_gene_id, chrom_subset )),]
  df$gene_length <- df$end_position - df$start_position
  
  temp_name <- paste0(chrom_subset_name, ": ", "gene_length density")
  ggl[[temp_name]] <- ggplot(df, aes(x=log10(gene_length)))+
    ggtitle(temp_name)+
    geom_density(alpha=0.5)
  
  temp_name <- paste0(chrom_subset_name, ": ", "gene_length barcode")
  ggl[[temp_name]] <- ggplot(df, aes(x=log10(gene_length),xend=log10(gene_length), y=0,yend=1))+
    ggtitle(temp_name)+
    geom_segment(alpha=0.5)+
    coord_cartesian(ylim = c(0.2,0.5), xlim = c(2, 6.5))+
    barcode_theme
  
  ## transcript length density + barcode plot
  df <- gene_sets$genes_all_transcript[which(is.element(gene_sets$genes_all_transcript$ensembl_gene_id, chrom_subset )),]
  df <- unique(df[,c("ensembl_gene_id", "transcript_length")])
  
  temp_name <- paste0(chrom_subset_name, ": ", "transcript_length density")
  ggl[[temp_name]] <- ggplot(df, aes(x=log10(transcript_length)))+
    ggtitle(temp_name)+
    geom_density(alpha=0.5)
  
  temp_name <- paste0(chrom_subset_name, ": ", "transcript_length barcode")
  ggl[[temp_name]] <-  ggplot(df, aes(x=log10(transcript_length),xend=log10(transcript_length), y=0,yend=1))+
    ggtitle(temp_name)+
    geom_segment(alpha=0.5)+
    coord_cartesian(ylim = c(0.2,0.5), xlim = c(2, 6.5))+
    barcode_theme
   
  
  
  ## exon length  plot
  df <- gene_sets$genes_all_transcript[which(is.element(gene_sets$genes_all_transcript$ensembl_gene_id, chrom_subset )),]
  #df <- gene_sets$genes_all_transcript[which(is.element(gene_sets$genes_all_transcript$ensembl_gene_id, sample(unique(gene_sets$genes_all$ensembl_gene_id),45) )),]
  df$exon_length <- df$exon_chrom_end - df$exon_chrom_start
  df$gene_factor_num <- as.numeric(as.factor(df$ensembl_gene_id))
  temp_name <- paste0(chrom_subset_name, ": ", "exon_length")
  ggl[[temp_name]] <- ggplot(df, aes(x=log10(exon_length),y=gene_factor_num))+
    ggtitle(temp_name)+
    geom_point(fill="grey20") +
    coord_cartesian( xlim = c(1, 4))+
    barcode_theme
  
  
  ## exon count plot
  temp_df <- as.data.frame(table(df$ensembl_transcript_id))
  colnames(temp_df)[1] <- "ensembl_transcript_id"
  colnames(temp_df)[2] <- "exon_count"
  df <- merge(df, temp_df, sort=FALSE)
  
  ## test 2d plot of exon count by min size
  
  temp <- split(df$exon_length, df$ensembl_transcript_id)
  temp <- sapply(temp, function(x) rank(-x, ties.method = "last"))
  df$exon_length_rank <- unsplit(temp, df$ensembl_transcript_id)
  
  temp <- split(df$exon_length, df$exon_length_rank) 
  temp <- sapply(temp, sort, decreasing=T)
  temp <- melt(temp)
  temp2 <- split(temp$value, temp$L1)
  num_genes <- length(unique(df$ensembl_gene_id))
  temp2 <- sapply(temp2, function(x) (1:length(x))/num_genes )
  temp2 <- unsplit(temp2, temp$L1)
  temp$percentile <- temp2
  temp$L1 <- as.numeric(temp$L1)
  temp <- temp[order(temp$L1, temp$percentile),]
  temp$L1 <- as.factor(temp$L1)
  colnames(temp) <- c("exon_length","exon_rank_by_length","percentile")
  temp_name <- paste0(chrom_subset_name, ": ", "genes_w_N_exons_of_size_or_greater")
  ggl[[temp_name]] <- ggplot(temp, aes(log10(exon_length), percentile, group=exon_rank_by_length, color=exon_rank_by_length))+
    ggtitle(temp_name)+
    geom_line()
  
  temp <- split(df$rank, df$ensembl_transcript_id)
  temp <-sapply(temp, function(x) order(-x))
  temp <- unsplit(temp, df$ensembl_transcript_id)
  df$rank_from_end <- temp
  temp_name <- paste0(chrom_subset_name, ": ", "exon_length v rank")
  ggl[[temp_name]] <- ggplot(df, aes(rank, log10(exon_length) ))+
    ggtitle(temp_name)+
    geom_point(alpha=0.5)
  temp_name <- paste0(chrom_subset_name, ": ", "exon_length v rank_from_end")
  ggl[[temp_name]] <- ggplot(df, aes(rank_from_end, log10(exon_length) ))+
    ggtitle(temp_name)+
    geom_point(alpha=0.5)
  
}

dirs$gene_arch_summary_figs <- paste0(dirs$figures,"gene_arch_summary_figures/")
if(!dir.exists(dirs$gene_arch_summary_figs)){dir.create(dirs$gene_arch_summary_figs)}
for(i in 1:length(ggl)){
  
  temp_filename <- paste0(dirs$gene_arch_summary_figs,names(ggl)[i],".png")
  temp_filename <- str_replace_all(temp_filename,"[ \t:]","_")
  w <- 7
  h <- ifelse(grepl("barcode", temp_filename), 3, 5)
  ggsave(temp_filename, ggl[[i]], device = "png", width=w, height=h)
}

```

```{r, look at basic disease association statistics }

### look at pLI

temp <- tempfile()
download.file("https://storage.googleapis.com/gcp-public-data--gnomad/release/2.1.1/constraint/gnomad.v2.1.1.lof_metrics.by_gene.txt.bgz",temp)
lof_metrics <- read.table( temp, skip=0, header = TRUE, sep = "\t")

temp <- tempfile()
download.file("http://geneontology.org/gene-associations/goa_human_complex.gaf.gz",temp)
go_complex_metrics <- read.table( temp, skip=0, header = TRUE, sep = "\t")

ggl2 <- list()
for(i in 14){ #  1:length(gene_sets$human_chromatin_modifiers$gene_subsets)){
  chrom_subset <- gene_sets$human_chromatin_modifiers$gene_subsets[[i]] 
  chrom_subset_name <- names(gene_sets$human_chromatin_modifiers$gene_subsets)[i]
  if(length(chrom_subset)==0){next}
  
  df <- lof_metrics[which(is.element(lof_metrics$gene_id, chrom_subset)),] 
  id_cols <- c("gene","gene_id","gene_type")
  var_cols <- c("oe_mis","oe_syn","oe_lof","no_lofs","obs_het_lof","obs_hom_lof", "pLI","oe_lof_upper","syn_z","mis_z","lof_z")
  df_melt <- melt(df, id.vars=id_cols, measure.vars = var_cols)
  
  temp_name <- paste0(chrom_subset_name, ": ", "gene_length density")
  ggplot(df_melt[grep("oe", df_melt$variable),],aes(x=variable, y=value))+
    geom_violin() 
  ggplot(df_melt[grep("_z", df_melt$variable),],aes(x=variable, y=value))+
    geom_violin()
  ggplot(df_melt[grep("pLI", df_melt$variable),],aes(y=value))+
    geom_histogram()
  ggplot(df_melt[grep("oe_lof_upper", df_melt$variable),],aes(y=value))+
    geom_histogram()
  
  
  ### GET all associated with Mendelian disorder (e.g. v pLI)

}
```

```{r, DRAFT extract protein complex associations}


### xml structure --- dissect xml in R to recapitulate each complex and it's member species
files$human_complexes <- paste0(dirs$temp_data,"human_complexes.zip")
#download.file("http://ftp.ebi.ac.uk/pub/databases/intact/complex/current/psi30/human.zip", files$human_complexes )
dirs$human_complexes <- str_remove(files$human_complexes,".zip")
#system(paste0("unzip ",files$human_complexes, " -d ", dirs$temp_data))
#system(paste0("mv ", dirs$temp_data, "human ", dirs$human_complexes))

temp_files <- list.files(dirs$human_complexes, full.names = T)
complex <- list()
i = 1
for(f in temp_files){
  if( i %% 20 ==1 ){print(paste0(basename(f),": ",i, " of ", length(temp_files)))}
  complex_metrics <- xmlParse(f)
  complex_metrics <- xmlToList(complex_metrics)
  i <- i+1
  
  complex_name <- complex_metrics$entry$interactionList$abstractInteraction$names$shortLabel
  complex_id <- paste0(complex_metrics$entry$interactionList$abstractInteraction$xref$primaryRef,collapse="__")
  complex[[complex_name]] <- list(name = complex_name,
                                  id = complex_id,
                                  members = c())
  for(interactor in complex_metrics$entry$interactorList){
    temp <- sapply(interactor$xref, function(x){paste0(x,collapse="__")} )
    temp <- temp[grep("ENSG.*gene",temp)]
    if(length(temp)==0){next}
    complex[[complex_name]]$members <- c(complex[[complex_name]]$members,
                                         str_extract(temp,"ENSG[0-9]*"))
  }  
}

### FOR EACH CHROM MODIFIER -> ID NUM OF COMPLEX SPECIES AND THEIR RESPECTIVE NUMBER OF TOTAL UNIQUE SUBUNITS ID'D
### PLOT SUMMARY DISTRIBUTION FOR ALL GENES
### BARCODE PLOT


# ggplot(df, aes(log2(transcript_length, )))+
#   geom_histogram(bins=500, alpha=0.7, fill="grey20")+
#   coord_cartesian(ylim = c(0.2,0.5), xlim = c(2, 6.5))+
#   barcode_theme
    #theme_bw()+


```

```{r, consider gene_families level analyses }
## https://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/csv/genefamily_db_tables/

```
