---
title: "general_summary_analyses"
author: "Leroy Bondhus"
date: "2023-07-10"
output: html_document
---

```{r, gene length barcode plot, pretty}
library(cowplot)

#### add to density plot line for all genes

df <- data.frame(gene_id=character(),
                 length = numeric(),
                 type = character(),
                 chrom_subset_name = character())

for(chrom_subset_name in c("control_set__non_chrom_mod","all_chrom_modifiers")){
  chrom_subset <- gene_sets$human_chromatin_modifiers$gene_subsets[[chrom_subset_name]]
  temp_df <- gene_sets$genes_all[which(is.element(gene_sets$genes_all$ensembl_gene_id, chrom_subset )),]
  df <- rbind(df, data.frame(gene_id=temp_df$ensembl_gene_id,
                             length=temp_df$end_position-temp_df$start_position,
                             type="gene",
                             chrom_subset_name=chrom_subset_name))
  
  temp_df <- gene_sets$genes_all_transcript[which(is.element(gene_sets$genes_all_transcript$ensembl_gene_id, chrom_subset ) &
                                                    gene_sets$genes_all_transcript$transcript_is_canonical),]
  temp_df <- unique(temp_df[,c("ensembl_gene_id","transcript_length")])
  df <- rbind(df, data.frame(gene_id=temp_df$ensembl_gene_id,
                             length=temp_df$transcript_length,
                             type="transcript",
                             chrom_subset_name=chrom_subset_name))
  
}


gg1 <- ggplot(df, aes(x=log10(length), color=paste0(chrom_subset_name,type) ))+
  scale_color_manual(values=c("#C77CFF","purple2","#00BFC4","turquoise3"))+
  scale_y_continuous(breaks=seq(0,1,by=0.1))+
  coord_cartesian(xlim = c(2, 6.2))+
  geom_step(aes(y=..y..),stat="ecdf")+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        legend.position = "none")

gg2 <- ggplot(df, aes(x=log10(length), color=paste0(chrom_subset_name,type) ))+
  scale_color_manual(values=c("#C77CFF","purple2","#00BFC4","turquoise3"))+
  coord_cartesian(xlim = c(2, 6.2))+
  geom_density(alpha=0.5)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        legend.position = "bottom")



filename <- paste0(dirs$figures, "gene_size.png")
ggsave(filename = filename,
       gridExtra::grid.arrange(gg1,gg2,nrow=2,heights=c(2,3)),
       height = 4, width=6)

```

```{r, first and last exon v all exons, pretty}
library(cowplot)

df <- data.frame(ensembl_gene_id = character(),
                 exon_length = character(),
                 exon_count = numeric(),
                 rank = numeric(),
                 rank_from_end = numeric(),
                 transcript_length = numeric(),
                 chrom_subset_name = character())
for(chrom_subset_name in c("control_set__non_chrom_mod","all_chrom_modifiers")){#}, "ubiq","non_ubiq")){
  ## get current subset of genes
  chrom_subset <- gene_sets$human_chromatin_modifiers$gene_subsets[[chrom_subset_name]] 
  if(length(chrom_subset)==0){next}
  ## map current subset to table with exon info
  
  ## exon length
  temp_df <- gene_sets$genes_all_transcript[which(is.element(gene_sets$genes_all_transcript$ensembl_gene_id,
                                                        chrom_subset )),]
  temp_df <- unique(temp_df)
  temp_df$exon_length <- temp_df$exon_chrom_end - temp_df$exon_chrom_start
  temp_df$gene_factor_num <- as.numeric(as.factor(temp_df$ensembl_gene_id))
  temp_name <- paste0(chrom_subset_name, ": ", "exon_length")
  
  ## exon count
  temp_df_2 <- as.data.frame(table(temp_df$ensembl_transcript_id))
  colnames(temp_df_2)[1] <- "ensembl_transcript_id"
  colnames(temp_df_2)[2] <- "exon_count"
  temp_df <- merge(temp_df, temp_df_2, sort=FALSE)
  
  ## exon rank from end
  temp_df$rank_from_end <- temp_df$exon_count - temp_df$rank + 1
  
  temp_df$chrom_subset_name <- chrom_subset_name
  df <- rbind(df, temp_df[,which(is.element(colnames(temp_df),colnames(df)))])
}


df$is_last <- df$rank_from_end == 1 & df$rank != 1
df$is_first <- df$rank == 1 & df$rank_from_end != 1

df$first_last_single_middle <- "middle"  
df$first_last_single_middle[which(df$exon_count==1)] <- "single"  
df$first_last_single_middle[which(df$is_first)] <- "first"
df$first_last_single_middle[which(df$is_last)] <- "last"
df$first_last_single_middle <- factor(df$first_last_single_middle, levels=c("first","middle","last","single"))
df$is_chrom <- !grepl("non_chrom",df$chrom_subset_name)

ggplot(df[df$first_last_single_middle != "single",], aes(x=log10(exon_length), linetype=chrom_subset_name))+
  coord_cartesian(xlim = c(0, 5.5))+
  geom_density(alpha=0.5)+
  theme_bw()+
  theme(axis.title.x = element_blank())+
  facet_grid(first_last_single_middle ~ .,scale="free")
filename <- paste0(dirs$figures, "exon_size__group_by_fmls.png")
ggsave(filename = filename, height=4,width=6)

ggplot(df[df$first_last_single_middle != "single",], aes(x=log10(transcript_length),y=exon_length/transcript_length,color=is_chrom))+
  coord_cartesian(xlim = c(2, 5))+
  scale_color_manual(values=c("#00BFC4","#C77CFF"))+
  geom_point(data=df[!df$is_chrom & df$first_last_single_middle != "single",],alpha=0.1,size=0.5)+
  geom_point(data=df[df$is_chrom & df$first_last_single_middle != "single",],alpha=0.7,size=0.5)+
  scale_y_continuous(breaks = seq(0,1,by=0.1))+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        panel.grid.minor.y = element_blank())+
  stat_smooth(method=lm)+
  facet_grid(first_last_single_middle ~ .,scale="free")
filename <- paste0(dirs$figures, "exon_size_v_transcript_len_group_by_fmls.png")
ggsave(filename = filename, height=6,width=9)


model <- lm(log10(exon_length) ~ 
              chrom_subset_name + log10(transcript_length),
            data = df[df$first_last_single_middle=="last",])
summary(model)


  
df$tau <- global_taus[df$ensembl_gene_id]
df$is_ubiq <- df$tau < 0.4
df <- df[complete.cases(df),]
temp_df <- unique(df[,c("ensembl_gene_id","exon_count","transcript_length","chrom_subset_name","is_chrom","is_ubiq","tau")])
ggplot(temp_df,aes(y=log2(exon_count), x=log10(transcript_length), color=is_chrom))+
  scale_color_manual(values=c("#00BFC4","#C77CFF"))+
  geom_point(data=df[which(df$first_last_single_middle=="last" & 
                             !df$is_chrom),], size=0.8, alpha=0.2)+
  geom_point(data=df[which(df$first_last_single_middle=="last" &
                             df$is_chrom),], size=1.6, alpha=0.8)+
  stat_smooth(method=lm)+
  #geom_density_2d()+
  xlim(2.5,4.5)

temp_df$tau <- global_taus[temp_df$ensembl_gene_id]
model <- lm(log2(exon_count) ~ 
              is_chrom + log10(transcript_length) + tau,
            data = temp_df)
summary(model)


ggplot(temp_df,aes(y=log2(exon_count), x=log10(transcript_length), color=paste0(is_ubiq,is_chrom) ))+
  scale_color_manual(values=c("gold2","#F8766D","#00BFC4","#C77CFF"))+
  theme_bw()+
  geom_jitter(height=0.1, data=df[which(df$first_last_single_middle=="last" &
                                          !df$is_chrom),],size=0.2, alpha=0.4)+
  geom_jitter(height=0.1,data=df[which(df$first_last_single_middle=="last" &
                             df$is_chrom),], size=0.6, alpha=0.8)+
  stat_smooth(method=lm)+
  xlim(2.5,4.5)+
  ylim(0.8,6.5)+
  facet_grid(. ~ is_ubiq )
filename <- paste0(dirs$figures, "exon_count_v_transcript_len.png")
ggsave(filename = filename, height=5,width=8)



```

```{r, exon counts}


```


```{r, disorder domain analyses}



protein_domains$is_chrom <- is.element(protein_domains$ensembl_gene_id,
                                      gene_sets$human_chromatin_modifiers$gene_subsets$all_chrom_modifiers)
protein_domains$is_chrom_f <- factor(protein_domains$is_chrom, levels=c("TRUE","FALSE"))
gg1 <- ggplot(protein_domains[protein_domains$prop_disordered > 0,], aes(y=log10(prop_disordered*Length), x=log10(Length),color=is_chrom_f ))+
  geom_abline(slope=1, intercept = log10(1-c(0,0.5,0.75,0.9,0.99)),color="grey40",linetype="dashed")+
  scale_color_manual(values=c("purple","steelblue3"))+
  geom_point(data=protein_domains[protein_domains$prop_disordered > 0 &
                                    !protein_domains$is_chrom,],color="#00BFC4",alpha=0.2,size=0.3)+
  geom_point(data=protein_domains[protein_domains$prop_disordered > 0 &
                                    protein_domains$is_chrom,],color="#C77CFF",alpha=0.6)+
  #geom_point(data=data.frame(x=log10(seq(0,1e4,length.out=100)),y=log10(0.1*seq(0,1e4,length.out=100))),aes(x=x,y=y),inherit.aes = F)+
  xlim(1,5)+ylim(1,4)+
  theme_bw()+
  stat_smooth(method=lm)+
  facet_grid(. ~ is_chrom_f)

gg2 <- ggplot(protein_domains[protein_domains$prop_disordered > 0,], aes(y=prop_disordered, x=log10(Length),color=is_chrom_f ))+
  scale_color_manual(values=c("purple","steelblue3"))+
  geom_abline(slope=0,intercept = 1-c(0,0.5,0.75,0.9,0.99),color="grey40",linetype="dashed")+
  geom_point(data=protein_domains[protein_domains$prop_disordered > 0 &
                                    !protein_domains$is_chrom,],color="#00BFC4",alpha=0.2,size=0.3)+
  geom_point(data=protein_domains[protein_domains$prop_disordered > 0 &
                                    protein_domains$is_chrom,],color="#C77CFF",alpha=0.6)+
  theme_bw()+
  scale_y_continuous(limits = c(0,1),breaks=seq(0,1,by=0.1))+
  scale_x_continuous(limits = c(1,4.5))+
  stat_smooth(method=lm)+
  facet_grid(. ~ is_chrom_f)


filename <- paste0(dirs$figures,"chrom_disordered_domain_proportion.png")
ggsave(filename = filename, gg1,
        #gridExtra::grid.arrange(gg1,gg2,nrow=2,heights=c(3,3)),
       height = 3, width=7)

model <- lm(log10(prop_disordered * Length) ~ 
              log10(Length) + is_chrom,
            data = protein_domains[protein_domains$prop_disordered > 0,])
summary(model)
model <- lm(log10(prop_disordered * Length) ~ 
              log10(Length) + is_chrom + is_chrom:log10(Length),
            data = protein_domains[protein_domains$prop_disordered > 0,])
summary(model)

```


```{r, DRAFT extract protein complex associations}
library(cowplot)

### xml structure --- dissect xml in R to recapitulate each complex and it's member species
files$human_complexes <- paste0(dirs$temp_data,"human_complexes.zip")
dirs$human_complexes <- str_remove(files$human_complexes,".zip")

## Note : complexes are curated for known interactions and function (?) so may be more restrictive definition than others
if(sum(grepl("\\.xml",list.files(dirs$human_complexes)))==0){
  download.file("http://ftp.ebi.ac.uk/pub/databases/intact/complex/current/psi30/human.zip", files$human_complexes )
  system(paste0("unzip ",files$human_complexes, " -d ", dirs$temp_data))
  system(paste0("mv ", dirs$temp_data, "human ", dirs$human_complexes))
}

temp_files <- list.files(dirs$human_complexes, full.names = T)
temp_files <- temp_files[grep("\\.xml",temp_files)]

complex <- list()
i = 1
for(f in temp_files){
  if( i %% 20 ==1 ){print(paste0(basename(f),": ",i, " of ", length(temp_files)))}
  complex_metrics <- xmlParse(f)
  complex_metrics <- xmlToList(complex_metrics)
  i <- i+1
  
  complex_name <- complex_metrics$entry$interactionList$abstractInteraction$names$shortLabel
  complex_id <- paste0(complex_metrics$entry$interactionList$abstractInteraction$xref$primaryRef,collapse="__")
  complex[[complex_name]] <- list(name = complex_name,
                                  id = complex_id,
                                  members = c())
  for(interactor in complex_metrics$entry$interactorList){
    temp <- sapply(interactor$xref, function(x){paste0(x,collapse="__")} )
    temp <- temp[grep("ENSG.*gene",temp)]
    if(length(temp)==0){next}
    complex[[complex_name]]$members <- c(complex[[complex_name]]$members,
                                         str_extract(temp,"ENSG[0-9]*"))
  }  
}

### FOR EACH CHROM MODIFIER -> ID NUM OF COMPLEX SPECIES AND THEIR RESPECTIVE NUMBER OF TOTAL UNIQUE SUBUNITS ID'D
### PLOT SUMMARY DISTRIBUTION FOR ALL GENES
### BARCODE PLOT

df <- data.frame(ensembl_gene_id = character(),
                 complex_name = character(),
                 complex_partners = character(),
                 complex_partner_count  = numeric(),
                 complex_count = numeric(),
                 chrom_subset_name = character())

for(i in 1:length(gene_sets$human_chromatin_modifiers$gene_subsets)){
  ## get current subset of genes
  chrom_subset <- gene_sets$human_chromatin_modifiers$gene_subsets[[i]]
  chrom_subset_name <- names(gene_sets$human_chromatin_modifiers$gene_subsets)[i]
  if(length(chrom_subset)==0){next}
  ## map current subset to table with exon info
  
  temp_list <- list(gene_id=chrom_subset)
  temp_list$complex_names <- vector(mode = "list", length=length(temp_list$gene_id))
  temp_list$complex_partners <- vector(mode = "list", length=length(temp_list$gene_id))
  temp_list$constitutive_partners <- vector(mode = "list", length=length(temp_list$gene_id))
    
  for(j in 1:length(complex)){
    temp_modifiers <- complex[[j]]$members[is.element(complex[[j]]$members, temp_list$gene_id)]
    if(length(temp_modifiers)==0){next}
    for(k in 1:length(temp_modifiers)){
      which <- which(temp_list$gene_id == temp_modifiers[k])
      temp_list$complex_names[[which]] <- c(temp_list$complex_names[[which]],complex[[j]]$name)
      temp_list$complex_partners[[which]] <- c(temp_list$complex_partners[[which]],list(complex[[j]]$members))
    }
  }
  for(j in 1:length(temp_list$complex_partners)){
    if(is.null(temp_list$complex_partners[[j]])){next}
    for(k in 1:length(temp_list$complex_partners[[j]])){
      if(k==1){
        tentative_constituitive <- temp_list$complex_partners[[j]][[k]]
      } else{
        tentative_constituitive <- intersect(tentative_constituitive, temp_list$complex_partners[[j]][[k]])
      }
    }
    temp_list$constitutive_partners[[j]] <- tentative_constituitive
  }
  if(all(sapply(temp_list$complex_names,function(x){all(is.null(x))}))){next}
  names(temp_list$complex_names) <- 
    names(temp_list$complex_partners) <- 
    names(temp_list$constitutive_partners) <-
    temp_list$gene_id
  # temp_list2 <- temp_list
  # temp_list2$complex_names <- NULL
  
  temp_list$complex_partners_concat <- sapply(temp_list$complex_partners,function(x){sapply(x, paste,collapse=",")})
  temp_list$complex_partners_count <- lapply(temp_list$complex_partners,function(x){sapply(x,length)})
  
 
  temp_df <- melt(temp_list$complex_names[!sapply(temp_list$complex_names,is.null)], value.name="complex_name")
  temp_df_2 <- melt(temp_list$complex_partners_concat[!sapply(temp_list$complex_names,is.null)], value.name="complex_partners")
  temp_df <- cbind(temp_df, temp_df_2[,1,drop=F])
  temp_df_2 <- melt(temp_list$complex_partners_count[!sapply(temp_list$complex_names,is.null)], value.name="complex_partner_count")
  temp_df <- cbind(temp_df, temp_df_2[,1,drop=F])
  colnames(temp_df)[which(colnames(temp_df)=="L1")] <- "ensembl_gene_id"
  temp_df <- cbind(temp_df, data.frame(complex_count=as.numeric(table(temp_df$ensembl_gene_id)[temp_df$ensembl_gene_id])))
  temp_df$chrom_subset_name <- chrom_subset_name
  df <- rbind(df, temp_df)
}


ggl <- list()
ggl[["complex_partner_count"]] <- ggplot(df, aes(x=complex_partner_count))+
  geom_histogram(bins=100, alpha=0.7, fill="grey20")+
  theme_classic()+
  theme(strip.text.y = element_text(angle = 0),
        panel.border = element_rect(fill=NA,color="grey50"))+
  facet_grid(chrom_subset_name ~ ., scales="free_y")

ggl[["complex_count"]] <- ggplot(unique(df[,c("ensembl_gene_id",
                                              "complex_count",
                                              "chrom_subset_name")]),
                                 aes(x=complex_count))+
  geom_histogram(bins=100, alpha=0.7, fill="grey20")+
  theme_classic()+
  theme(strip.text.y = element_text(angle = 0),
        panel.border = element_rect(fill=NA,color="grey50"))+
  facet_grid(chrom_subset_name ~ ., scales="free_y")
  

filename <- paste0(dirs$figures, "complex_partner_count__partition_by_chromModType.png")
ggsave(filename, ggl$complex_partner_count, height = 6, width =8, units="in" )
filename <- paste0(dirs$figures, "complex_count__partition_by_chromModType.png")
ggsave(filename, ggl$complex_count, height = 6, width =8, units="in")
```

```{r, paralog groups analysis}
### number of paralogs per chrom mod
### number of paralogs groups
# for(i in 1:length(gene_sets$human_chromatin_modifiers$gene_subsets)){
#   ## get current subset of genes
#   chrom_subset <- gene_sets$human_chromatin_modifiers$gene_subsets[[i]] 
#   chrom_subset_name <- names(gene_sets$human_chromatin_modifiers$gene_subsets)[i]
#   if(length(chrom_subset)==0){next}
#   df <- data.frame(id=character(),
#                    num_paralogs=numeric(),
#                    paralog_group=character())
# 
#   
#   
# }

### divergence points of paralogs ## consider representating in a tree

#### want for downstream look at phenotype similarity between homologs

## also ## look at constitutive partner phenotype and expression similarity 
```


```{r, consider gene_families level analyses }
## https://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/csv/genefamily_db_tables/

```


```{r}

```
