---
title: "general_summary_analyses"
author: "Leroy Bondhus"
date: "2023-07-10"
output: html_document
---

```{r, gene length barcode plot, pretty}
library(cowplot)

#### add to density plot line for all genes

df <- cbind(gene_sets$genes_all[0,],
            data.frame(gene_length = numeric(), chrom_subset_name = character()))
for(i in 1:length(gene_sets$human_chromatin_modifiers$gene_subsets)){
  ## get current subset of genes
  chrom_subset <- gene_sets$human_chromatin_modifiers$gene_subsets[[i]] 
  chrom_subset_name <- names(gene_sets$human_chromatin_modifiers$gene_subsets)[i]
  if(length(chrom_subset)==0){next}
  ## map current subset to table with gene lenght info
  temp_df <- gene_sets$genes_all[which(is.element(gene_sets$genes_all$ensembl_gene_id, chrom_subset )),]
  temp_df$gene_length <- temp_df$end_position - temp_df$start_position
  temp_df$chrom_subset_name <- chrom_subset_name
  df <- rbind(df, temp_df)
}

ggl <- list()
ggl[["all_density"]] <- ggplot(df[grep("^all_|^control",df$chrom_subset_name),], aes(x=log10(gene_length), color=chrom_subset_name))+
  coord_cartesian(xlim = c(2, 6.5))+
  geom_density(alpha=0.5)+
  barcode_theme+
  theme(axis.title.x = element_blank())

ggl[["barcodes"]] <- ggplot(df[grep("^control",df$chrom_subset_name,invert=T),], aes(x=log10(gene_length),xend=log10(gene_length), y=0,yend=1))+
  geom_segment(alpha=0.5)+
  coord_cartesian(ylim = c(0.2,0.5), xlim = c(2, 6.5))+
  barcode_theme+
  facet_grid(chrom_subset_name ~ . )+
  theme(strip.text.y = element_text(angle = 0))
  
ggl[["composite"]] <- plot_grid(ggl$all_density, ggl$barcodes, align="v",axis="r",nrow = 2, rel_heights = c(3,8))
ggl$composite

filename <- paste0(dirs$figures, "gene_size__partition_by_chromModType.png")
save_plot(filename = filename, ggl$composite, base_height = 6, base_asp = 1.2)
```

```{r, transcript length barcode plot, pretty}
library(cowplot)

#### add to density plot line for all genes

df <- data.frame(ensembl_gene_id = character(),
                 transcript_length = numeric(),
                 chrom_subset_name = character())
for(i in 1:length(gene_sets$human_chromatin_modifiers$gene_subsets)){
  ## get current subset of genes
  chrom_subset <- gene_sets$human_chromatin_modifiers$gene_subsets[[i]] 
  chrom_subset_name <- names(gene_sets$human_chromatin_modifiers$gene_subsets)[i]
  if(length(chrom_subset)==0){next}
  ## map current subset to table with transcript length info
  temp_df <- gene_sets$genes_all_transcript[which(is.element(gene_sets$genes_all_transcript$ensembl_gene_id, chrom_subset )),]
  temp_df <- unique(temp_df[,c("ensembl_gene_id", "transcript_length")])
  temp_df$chrom_subset_name <- chrom_subset_name
  df <- rbind(df, temp_df)
}

ggl <- list()
ggl[["all_density"]] <- ggplot(df[grep("^all_|^control",df$chrom_subset_name),], aes(x=log10(transcript_length),color=chrom_subset_name))+
  coord_cartesian(xlim = c(2, 6.5))+
  geom_density(alpha=0.5)+
  barcode_theme+
  theme(axis.title.x = element_blank())

ggl[["barcodes"]] <- ggplot(df[grep("^control",df$chrom_subset_name,invert=T),], aes(x=log10(transcript_length),xend=log10(transcript_length), y=0,yend=1))+
  geom_segment(alpha=0.5)+
  coord_cartesian(ylim = c(0.2,0.5), xlim = c(2, 6.5))+
  barcode_theme+
  facet_grid(chrom_subset_name ~ . )+
  theme(strip.text.y = element_text(angle = 0))
  
ggl[["composite"]] <- plot_grid(ggl$all_density, ggl$barcodes, align="v",axis="r",nrow = 2, rel_heights = c(3,8))
ggl$composite

filename <- paste0(dirs$figures, "transcript_size__partition_by_chromModType.png")
save_plot(filename = filename, ggl$composite,base_height = 6, base_asp = 1.2)
```

```{r, first and last exon v all exons, pretty}
library(cowplot)

df <- data.frame(ensembl_gene_id = character(),
                 exon_length = character(),
                 exon_count = numeric(),
                 rank = numeric(),
                 rank_from_end = numeric(),
                 chrom_subset_name = character())
for(i in 1:length(gene_sets$human_chromatin_modifiers$gene_subsets)){
  ## get current subset of genes
  chrom_subset <- gene_sets$human_chromatin_modifiers$gene_subsets[[i]] 
  chrom_subset_name <- names(gene_sets$human_chromatin_modifiers$gene_subsets)[i]
  if(length(chrom_subset)==0){next}
  ## map current subset to table with exon info
  
  ## exon length
  temp_df <- gene_sets$genes_all_transcript[which(is.element(gene_sets$genes_all_transcript$ensembl_gene_id,
                                                        chrom_subset )),]
  temp_df$exon_length <- temp_df$exon_chrom_end - temp_df$exon_chrom_start
  temp_df$gene_factor_num <- as.numeric(as.factor(temp_df$ensembl_gene_id))
  temp_name <- paste0(chrom_subset_name, ": ", "exon_length")
  
  ## exon count
  temp_df_2 <- as.data.frame(table(temp_df$ensembl_transcript_id))
  colnames(temp_df_2)[1] <- "ensembl_transcript_id"
  colnames(temp_df_2)[2] <- "exon_count"
  temp_df <- merge(temp_df, temp_df_2, sort=FALSE)
  
  ## exon rank from end
  temp_df$rank_from_end <- temp_df$exon_count - temp_df$rank + 1
  
  temp_df$chrom_subset_name <- chrom_subset_name
  df <- rbind(df, temp_df[,which(is.element(colnames(temp_df),colnames(df)))])
}

ggl <- list()
df$is_last <- df$rank_from_end == 1 & df$rank != 1
df$is_first <- df$rank == 1 & df$rank_from_end != 1

df$first_last_single_middle <- "middle"  
df$first_last_single_middle[which(df$exon_count==1)] <- "single"  
df$first_last_single_middle[which(df$is_first)] <- "first"
df$first_last_single_middle[which(df$is_last)] <- "last"

ggl[["all_density"]] <- ggplot(df[grep("^all_|^control",df$chrom_subset_name),], aes(x=log10(exon_length),
                                                                                     color=first_last_single_middle,
                                                                                     linetype=chrom_subset_name))+
  coord_cartesian(xlim = c(0, 5.5))+
  geom_density(alpha=0.5)+
  barcode_theme+
  theme(axis.title.x = element_blank())+
  facet_grid(first_last_single_middle ~  .)
table(df$first_last_single_middle)

ggl[["barcodes"]] <- ggplot(df[grep("^control",df$chrom_subset_name, invert=T),], aes(x=log10(exon_length),xend=log10(exon_length),
               y=0.25*as.numeric(as.factor(first_last_single_middle)),
               yend=0.25*as.numeric(as.factor(first_last_single_middle))+0.25,
               color=first_last_single_middle))+
  geom_segment(alpha=0.5)+
  coord_cartesian(ylim = c(0.25,1.25), xlim = c(0, 5.5))+
  barcode_theme+
  facet_grid(chrom_subset_name ~ . )+
  theme(strip.text.y = element_text(angle = 0))
  
ggl[["composite"]] <- plot_grid(ggl$all_density, ggl$barcodes, align="v",axis="r",nrow = 2, rel_heights = c(6,8))
ggl$composite

filename <- paste0(dirs$figures, "exon_size__group_by_fmls__partition_by_chromModType.png")
save_plot(filename = filename, ggl$composite, base_height = 6, base_asp = 1.2)


# ggl[["all_density"]] <- ggplot(df[grep("^all_|^control",df$chrom_subset_name),], aes(x=log10(exon_length), color=chrom_subset_name))+
#   coord_cartesian(xlim = c(0, 5.5))+
#   geom_density(alpha=0.5)+
#   barcode_theme+
#   theme(axis.title.x = element_blank())+
#   facet_grid(first_last_single_middle ~ . )
# table(df$first_last_single_middle)
```


```{r, DRAFT extract protein complex associations}
library(cowplot)

### xml structure --- dissect xml in R to recapitulate each complex and it's member species
files$human_complexes <- paste0(dirs$temp_data,"human_complexes.zip")
dirs$human_complexes <- str_remove(files$human_complexes,".zip")

## Note : complexes are curated for known interactions and function (?) so may be more restrictive definition than others
if(sum(grepl("\\.xml",list.files(dirs$human_complexes)))==0){
  download.file("http://ftp.ebi.ac.uk/pub/databases/intact/complex/current/psi30/human.zip", files$human_complexes )
  system(paste0("unzip ",files$human_complexes, " -d ", dirs$temp_data))
  system(paste0("mv ", dirs$temp_data, "human ", dirs$human_complexes))
}

temp_files <- list.files(dirs$human_complexes, full.names = T)
temp_files <- temp_files[grep("\\.xml",temp_files)]

complex <- list()
i = 1
for(f in temp_files){
  if( i %% 20 ==1 ){print(paste0(basename(f),": ",i, " of ", length(temp_files)))}
  complex_metrics <- xmlParse(f)
  complex_metrics <- xmlToList(complex_metrics)
  i <- i+1
  
  complex_name <- complex_metrics$entry$interactionList$abstractInteraction$names$shortLabel
  complex_id <- paste0(complex_metrics$entry$interactionList$abstractInteraction$xref$primaryRef,collapse="__")
  complex[[complex_name]] <- list(name = complex_name,
                                  id = complex_id,
                                  members = c())
  for(interactor in complex_metrics$entry$interactorList){
    temp <- sapply(interactor$xref, function(x){paste0(x,collapse="__")} )
    temp <- temp[grep("ENSG.*gene",temp)]
    if(length(temp)==0){next}
    complex[[complex_name]]$members <- c(complex[[complex_name]]$members,
                                         str_extract(temp,"ENSG[0-9]*"))
  }  
}

### FOR EACH CHROM MODIFIER -> ID NUM OF COMPLEX SPECIES AND THEIR RESPECTIVE NUMBER OF TOTAL UNIQUE SUBUNITS ID'D
### PLOT SUMMARY DISTRIBUTION FOR ALL GENES
### BARCODE PLOT

df <- data.frame(ensembl_gene_id = character(),
                 complex_name = character(),
                 complex_partners = character(),
                 complex_partner_count  = numeric(),
                 complex_count = numeric(),
                 chrom_subset_name = character())

for(i in 1:length(gene_sets$human_chromatin_modifiers$gene_subsets)){
  ## get current subset of genes
  chrom_subset <- gene_sets$human_chromatin_modifiers$gene_subsets[[i]]
  chrom_subset_name <- names(gene_sets$human_chromatin_modifiers$gene_subsets)[i]
  if(length(chrom_subset)==0){next}
  ## map current subset to table with exon info
  
  temp_list <- list(gene_id=chrom_subset)
  temp_list$complex_names <- vector(mode = "list", length=length(temp_list$gene_id))
  temp_list$complex_partners <- vector(mode = "list", length=length(temp_list$gene_id))
  temp_list$constitutive_partners <- vector(mode = "list", length=length(temp_list$gene_id))
    
  for(j in 1:length(complex)){
    temp_modifiers <- complex[[j]]$members[is.element(complex[[j]]$members, temp_list$gene_id)]
    if(length(temp_modifiers)==0){next}
    for(k in 1:length(temp_modifiers)){
      which <- which(temp_list$gene_id == temp_modifiers[k])
      temp_list$complex_names[[which]] <- c(temp_list$complex_names[[which]],complex[[j]]$name)
      temp_list$complex_partners[[which]] <- c(temp_list$complex_partners[[which]],list(complex[[j]]$members))
    }
  }
  for(j in 1:length(temp_list$complex_partners)){
    if(is.null(temp_list$complex_partners[[j]])){next}
    for(k in 1:length(temp_list$complex_partners[[j]])){
      if(k==1){
        tentative_constituitive <- temp_list$complex_partners[[j]][[k]]
      } else{
        tentative_constituitive <- intersect(tentative_constituitive, temp_list$complex_partners[[j]][[k]])
      }
    }
    temp_list$constitutive_partners[[j]] <- tentative_constituitive
  }
  if(all(sapply(temp_list$complex_names,function(x){all(is.null(x))}))){next}
  names(temp_list$complex_names) <- 
    names(temp_list$complex_partners) <- 
    names(temp_list$constitutive_partners) <-
    temp_list$gene_id
  # temp_list2 <- temp_list
  # temp_list2$complex_names <- NULL
  
  temp_list$complex_partners_concat <- sapply(temp_list$complex_partners,function(x){sapply(x, paste,collapse=",")})
  temp_list$complex_partners_count <- lapply(temp_list$complex_partners,function(x){sapply(x,length)})
  
 
  temp_df <- melt(temp_list$complex_names[!sapply(temp_list$complex_names,is.null)], value.name="complex_name")
  temp_df_2 <- melt(temp_list$complex_partners_concat[!sapply(temp_list$complex_names,is.null)], value.name="complex_partners")
  temp_df <- cbind(temp_df, temp_df_2[,1,drop=F])
  temp_df_2 <- melt(temp_list$complex_partners_count[!sapply(temp_list$complex_names,is.null)], value.name="complex_partner_count")
  temp_df <- cbind(temp_df, temp_df_2[,1,drop=F])
  colnames(temp_df)[which(colnames(temp_df)=="L1")] <- "ensembl_gene_id"
  temp_df <- cbind(temp_df, data.frame(complex_count=as.numeric(table(temp_df$ensembl_gene_id)[temp_df$ensembl_gene_id])))
  temp_df$chrom_subset_name <- chrom_subset_name
  df <- rbind(df, temp_df)
}


ggl <- list()
ggl[["complex_partner_count"]] <- ggplot(df, aes(x=complex_partner_count))+
  geom_histogram(bins=100, alpha=0.7, fill="grey20")+
  theme_classic()+
  theme(strip.text.y = element_text(angle = 0),
        panel.border = element_rect(fill=NA,color="grey50"))+
  facet_grid(chrom_subset_name ~ ., scales="free_y")

ggl[["complex_count"]] <- ggplot(unique(df[,c("ensembl_gene_id",
                                              "complex_count",
                                              "chrom_subset_name")]),
                                 aes(x=complex_count))+
  geom_histogram(bins=100, alpha=0.7, fill="grey20")+
  theme_classic()+
  theme(strip.text.y = element_text(angle = 0),
        panel.border = element_rect(fill=NA,color="grey50"))+
  facet_grid(chrom_subset_name ~ ., scales="free_y")
  

filename <- paste0(dirs$figures, "complex_partner_count__partition_by_chromModType.png")
ggsave(filename, ggl$complex_partner_count, height = 6, width =8, units="in" )
filename <- paste0(dirs$figures, "complex_count__partition_by_chromModType.png")
ggsave(filename, ggl$complex_count, height = 6, width =8, units="in")
```

```{r, paralog groups analysis}
### number of paralogs per chrom mod
### number of paralogs groups
# for(i in 1:length(gene_sets$human_chromatin_modifiers$gene_subsets)){
#   ## get current subset of genes
#   chrom_subset <- gene_sets$human_chromatin_modifiers$gene_subsets[[i]] 
#   chrom_subset_name <- names(gene_sets$human_chromatin_modifiers$gene_subsets)[i]
#   if(length(chrom_subset)==0){next}
#   df <- data.frame(id=character(),
#                    num_paralogs=numeric(),
#                    paralog_group=character())
# 
#   
#   
# }

### divergence points of paralogs ## consider representating in a tree

#### want for downstream look at phenotype similarity between homologs

## also ## look at constitutive partner phenotype and expression similarity 
```


```{r, consider gene_families level analyses }
## https://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/csv/genefamily_db_tables/

```


```{r}

```
